{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import shlex\n",
    "pd.set_option('display.max_rows', 75)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Thoughts\n",
    "Can a contig can still be a contaminant even if it is at lower relative coverage in a control than in the sample?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sample_name, contigs db, and profile db, generate needed input files\n",
    "# also take as input, a list of controls sample names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_name = 'AWTP_2_RO2_bulk_3'\n",
    "#contig_data_file = '/Users/rosekantor/data/awtp2_metagenomics/decontamination/KNLK_58_contigs_basic_info.txt'\n",
    "taxonomy_file = '/Users/rosekantor/data/awtp2_metagenomics/test_decontamination/KNLK_58_taxonomy_splits.txt'\n",
    "cov_means_file = '/Users/rosekantor/data/awtp2_metagenomics/test_decontamination/KNLK_58_filt_mean_coverage_Q2Q3_contigs.txt'\n",
    "cov_stds_file = '/Users/rosekantor/data/awtp2_metagenomics/test_decontamination/KNLK_58_filt_std_coverage_contigs.txt'\n",
    "detect_file = '/Users/rosekantor/data/awtp2_metagenomics/test_decontamination/KNLK_58_filt_detection_contigs.txt'\n",
    "out_file = '/Users/rosekantor/data/awtp2_metagenomics/test_decontamination/KNLK_58_contaminants/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_table(anvio_file):\n",
    "    # import coverage means table\n",
    "    df = pd.read_csv(anvio_file, sep='\\t')\n",
    "    #df = df.drop(columns = 'contig') #first col is actually the split so remove it since we want contigs (luckily values for all splits in a contig are identical since we specified contig)\n",
    "    df = df.rename(columns = {'contig': 'split', '__parent__' : 'contig', sample_name : 's'}) \n",
    "    #df = df.drop_duplicates() # collapse down to just contig level\n",
    "    df = df.drop(columns = 'contig')\n",
    "    df = df.set_index('split')\n",
    "\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "def score_contigs_cov(cov_means, cov_stds):\n",
    "    '''Check whether the contig mean coverage is greater in sample than control and\n",
    "    whether contig normalized standard deviation of coverage is less in sample than in control.\n",
    "    If both are true, then the contig is not a contaminant.\n",
    "    If one is false, contig will be in the returned dataframe for further analysis.'''\n",
    "    \n",
    "    # calculate relative coverage\n",
    "    cov_means_norm = cov_means.div(cov_means.sum(axis=0), axis=1)\n",
    "    # compare each column to s and make a matrix that is boolean control < sample (True is noncontam contig)\n",
    "    cov_means_norm_lt = cov_means_norm.lt(cov_means_norm.s, axis=0)\n",
    "\n",
    "    # get coefficient of variance (mean-normalized standard deviation of coverage for each contig)\n",
    "    cov_stds_norm = cov_stds / cov_means\n",
    "    cov_stds_norm = cov_stds_norm.replace([np.nan, np.inf], 0)\n",
    "    # compare each column to s and make a matrix that is boolean control > sample (True is noncontam contig)\n",
    "    cov_stds_norm_gt = cov_stds_norm.gt(cov_stds_norm.s, axis=0)\n",
    "\n",
    "    ## combinations of cov_means_norm_lt and cov_stds_norm_gt\n",
    "    # convert boolean to number and then add, so that:\n",
    "    # True & True = 2 (noncontam)\n",
    "    # True & False, False & True = 1 (unknown)\n",
    "    # False & False = 0 (contam)\n",
    "    \n",
    "    iscontam = cov_means_norm_lt*1 + cov_stds_norm_gt*1\n",
    "\n",
    "    ## special cases of False&True and True&False (both have value of 1)\n",
    "\n",
    "    ## for any contig where (cov_means_norm.s == 0), change iscontam to 0 for all controls\n",
    "    s_no_cov = (cov_means_norm.s == 0) * 1\n",
    "    iscontam = iscontam.subtract(s_no_cov, axis=0) # subtract 1 in cases where sample has no cov.\n",
    "\n",
    "    # for any contig x control where (cov_means_norm.c == 0), change iscontam to 2 for that control\n",
    "    c_no_cov = (cov_means_norm==0 * 1)\n",
    "    iscontam = iscontam.add(c_no_cov) # add 1 in cases where control has no cov.\n",
    "    \n",
    "    iscontam = iscontam.drop(columns = 's')\n",
    "    \n",
    "    # Remove all contigs where every control comparison yielded a 2 (noncontam)\n",
    "    cov_contams = iscontam[~((iscontam == 2).all(axis = 'columns'))]\n",
    "\n",
    "    return(cov_contams)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def score_contigs_detection(detection, cov_contams):\n",
    "    '''Check detection on suspected contaminant contigs in all controls:\n",
    "    if detection in control is less than detection in sample - 0.05 then remove from contaminant list\n",
    "    because coverage is probably due to highly conserved sequence or mismapping, not true presence in sample'''\n",
    "\n",
    "    # subset detection consider to just possible contaminant contigs\n",
    "    detect_posscontam = detection[detection.index.isin(cov_contams.index)]\n",
    "\n",
    "    # make masking df where:\n",
    "    # True means detection in control is not within 0.05 of the detection in sample, so it is likely to be real\n",
    "    # False means detection in control is similar to detection in sample, so it is likely to be a contaminant\n",
    "    detect_posscontam_lt = detect_posscontam.lt(detect_posscontam.s - 0.05, axis=0)\n",
    "    detect_posscontam_lt = detect_posscontam_lt.drop(columns = 's')\n",
    "    final_contams = cov_contams[detect_posscontam_lt==False]\n",
    "    final_contams = final_contams[~((final_contams.isna()).all(axis = 'columns'))]\n",
    "    \n",
    "    return(final_contams)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "def taxonomy_remove(tax_names, taxonomy_file):\n",
    "    tax = pd.read_csv(taxonomy_file, sep = '\\t', names=['split', 'genus'], index_col=0)\n",
    "    contam_splits = pd.DataFrame(tax[tax.genus.isin(tax_names)].index)\n",
    "    return(contam_splits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run functions\n",
    "tax = pd.read_csv(taxonomy_file, sep = '\\t', names=['split', 'genus'], index_col=0)\n",
    "\n",
    "cov_means = import_table(cov_means_file)\n",
    "cov_stds = import_table(cov_stds_file)\n",
    "detection = import_table(detect_file)\n",
    "cov_contams = score_contigs_cov(cov_means, cov_stds)\n",
    "detect_contams = score_contigs_detection(detection, cov_contams)\n",
    "#possible_contams.to_csv('~/data/awtp2_metagenomics/KNLK_58_contaminants/contam/KNLK_58_contam_scores.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>split</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>KNLK_58_123223_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>KNLK_58_100836_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KNLK_58_481724_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>KNLK_58_33625_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KNLK_58_11221_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>946</td>\n",
       "      <td>KNLK_58_90410_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>947</td>\n",
       "      <td>KNLK_58_91291_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>949</td>\n",
       "      <td>KNLK_58_93484_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>950</td>\n",
       "      <td>KNLK_58_93849_split_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>954</td>\n",
       "      <td>KNLK_58_96417_split_00001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>992 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          split\n",
       "0    KNLK_58_123223_split_00001\n",
       "1    KNLK_58_100836_split_00001\n",
       "2    KNLK_58_481724_split_00001\n",
       "3     KNLK_58_33625_split_00001\n",
       "4     KNLK_58_11221_split_00001\n",
       "..                          ...\n",
       "946   KNLK_58_90410_split_00001\n",
       "947   KNLK_58_91291_split_00001\n",
       "949   KNLK_58_93484_split_00001\n",
       "950   KNLK_58_93849_split_00001\n",
       "954   KNLK_58_96417_split_00001\n",
       "\n",
       "[992 rows x 1 columns]"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = taxonomy_remove(['Lysobacter', 'Asaccharospora'], taxonomy_file)\n",
    "d = pd.DataFrame(detect_contams.index)\n",
    "pd.concat([t, d]).drop_duplicates('split')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#iscontam = iscontam.add_prefix('score_')\n",
    "# iscontam = iscontam.replace({0: 'contam', 1: 'unknown', 2: 'noncontam'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(965, 6)"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detect_contams.shape#[(detect_contams == 2).any(axis = 'columns')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1033"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possible_contams_contigs = possible_contams.index.to_list()\n",
    "len(possible_contams_contigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "970"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find contigs that had higher or nearly as high detection in any blank as in sample\n",
    "control_names = detection.columns.to_list()\n",
    "control_names.remove('s') # get list of controls to use to subset the detect dataframe\n",
    "contam_by_detect_df = detect[(detect[control_names].max(axis=1) + 0.05) >= detect.s]\n",
    "detect_contam_contigs = contam_by_detect_df.index.to_list()\n",
    "len(detect_contam_contigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(set(possible_contams_contigs) - set(detect_contam_contigs)) # = 68\n",
    "detect_not_cov = list(set(detect_contam_contigs) - set(cov_contams.index)) # = 5\n",
    "\n",
    "#not_detect = list(set(possible_contams_contigs) - set(detect_contam_contigs))\n",
    "#not_detect_df = detect[detect.index.isin(not_detect)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_91778_split_00001</td>\n",
       "      <td>Mycobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_51681_split_00001</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_175146_split_00001</td>\n",
       "      <td>T5virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_387596_split_00001</td>\n",
       "      <td>Unknown_Caudovirales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_433385_split_00001</td>\n",
       "      <td>Unknown_Verrucomicrobia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              genus\n",
       "split                                              \n",
       "KNLK_58_91778_split_00001             Mycobacterium\n",
       "KNLK_58_51681_split_00001                       NaN\n",
       "KNLK_58_175146_split_00001                  T5virus\n",
       "KNLK_58_387596_split_00001     Unknown_Caudovirales\n",
       "KNLK_58_433385_split_00001  Unknown_Verrucomicrobia"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax[tax.index.isin(detect_not_cov)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "965"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1033 - 68"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'KNLK_58_10080_split_00001' in not_detect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s</th>\n",
       "      <th>neg_ex_1</th>\n",
       "      <th>neg_field_1</th>\n",
       "      <th>neg_field_2</th>\n",
       "      <th>neg_field_bf</th>\n",
       "      <th>pos_ex_1</th>\n",
       "      <th>pos_ex_2</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_175146_split_00001</td>\n",
       "      <td>161.747617</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.593578</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.660813</td>\n",
       "      <td>0.242348</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T5virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_387596_split_00001</td>\n",
       "      <td>82.869547</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.001223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.252752</td>\n",
       "      <td>0.234815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Unknown_Caudovirales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_433385_split_00001</td>\n",
       "      <td>62.796555</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.543068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.739130</td>\n",
       "      <td>0.428220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Unknown_Verrucomicrobia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_51681_split_00001</td>\n",
       "      <td>18.073084</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.103387</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.178253</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_91778_split_00001</td>\n",
       "      <td>79.135472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.890374</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.396613</td>\n",
       "      <td>0.229947</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Mycobacterium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     s  neg_ex_1  neg_field_1  neg_field_2  \\\n",
       "split                                                                        \n",
       "KNLK_58_175146_split_00001  161.747617       0.0     7.593578          0.0   \n",
       "KNLK_58_387596_split_00001   82.869547       0.0     4.001223          0.0   \n",
       "KNLK_58_433385_split_00001   62.796555       0.0     4.543068          0.0   \n",
       "KNLK_58_51681_split_00001    18.073084       0.0     5.103387          0.0   \n",
       "KNLK_58_91778_split_00001    79.135472       0.0     6.890374          0.0   \n",
       "\n",
       "                            neg_field_bf  pos_ex_1  pos_ex_2  \\\n",
       "split                                                          \n",
       "KNLK_58_175146_split_00001     12.660813  0.242348       0.0   \n",
       "KNLK_58_387596_split_00001      8.252752  0.234815       0.0   \n",
       "KNLK_58_433385_split_00001      6.739130  0.428220       0.0   \n",
       "KNLK_58_51681_split_00001       8.178253  0.000000       0.0   \n",
       "KNLK_58_91778_split_00001       6.396613  0.229947       0.0   \n",
       "\n",
       "                                              genus  \n",
       "split                                                \n",
       "KNLK_58_175146_split_00001                  T5virus  \n",
       "KNLK_58_387596_split_00001     Unknown_Caudovirales  \n",
       "KNLK_58_433385_split_00001  Unknown_Verrucomicrobia  \n",
       "KNLK_58_51681_split_00001                       NaN  \n",
       "KNLK_58_91778_split_00001             Mycobacterium  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s</th>\n",
       "      <th>neg_ex_1</th>\n",
       "      <th>neg_field_1</th>\n",
       "      <th>neg_field_2</th>\n",
       "      <th>neg_field_bf</th>\n",
       "      <th>pos_ex_1</th>\n",
       "      <th>pos_ex_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_175146_split_00001</td>\n",
       "      <td>54.527095</td>\n",
       "      <td>0.378832</td>\n",
       "      <td>4.048214</td>\n",
       "      <td>0.190373</td>\n",
       "      <td>7.468798</td>\n",
       "      <td>1.501851</td>\n",
       "      <td>0.498211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_387596_split_00001</td>\n",
       "      <td>21.544431</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.218208</td>\n",
       "      <td>0.408411</td>\n",
       "      <td>6.105507</td>\n",
       "      <td>1.021115</td>\n",
       "      <td>0.172214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_433385_split_00001</td>\n",
       "      <td>32.034552</td>\n",
       "      <td>0.375200</td>\n",
       "      <td>3.448368</td>\n",
       "      <td>0.408702</td>\n",
       "      <td>3.608583</td>\n",
       "      <td>1.393819</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_51681_split_00001</td>\n",
       "      <td>10.508541</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.019266</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.913403</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_91778_split_00001</td>\n",
       "      <td>28.199513</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.917468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.425999</td>\n",
       "      <td>1.419072</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    s  neg_ex_1  neg_field_1  neg_field_2  \\\n",
       "split                                                                       \n",
       "KNLK_58_175146_split_00001  54.527095  0.378832     4.048214     0.190373   \n",
       "KNLK_58_387596_split_00001  21.544431  0.000000     3.218208     0.408411   \n",
       "KNLK_58_433385_split_00001  32.034552  0.375200     3.448368     0.408702   \n",
       "KNLK_58_51681_split_00001   10.508541  0.000000     3.019266     0.000000   \n",
       "KNLK_58_91778_split_00001   28.199513  0.000000     3.917468     0.000000   \n",
       "\n",
       "                            neg_field_bf  pos_ex_1  pos_ex_2  \n",
       "split                                                         \n",
       "KNLK_58_175146_split_00001      7.468798  1.501851  0.498211  \n",
       "KNLK_58_387596_split_00001      6.105507  1.021115  0.172214  \n",
       "KNLK_58_433385_split_00001      3.608583  1.393819  0.000000  \n",
       "KNLK_58_51681_split_00001       4.913403  0.000000  0.000000  \n",
       "KNLK_58_91778_split_00001       5.425999  1.419072  0.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s</th>\n",
       "      <th>neg_ex_1</th>\n",
       "      <th>neg_field_1</th>\n",
       "      <th>neg_field_2</th>\n",
       "      <th>neg_field_bf</th>\n",
       "      <th>pos_ex_1</th>\n",
       "      <th>pos_ex_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_175146_split_00001</td>\n",
       "      <td>0.997489</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.954306</td>\n",
       "      <td>0.037660</td>\n",
       "      <td>0.974140</td>\n",
       "      <td>0.371077</td>\n",
       "      <td>0.189305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_387596_split_00001</td>\n",
       "      <td>0.998368</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.917805</td>\n",
       "      <td>0.092392</td>\n",
       "      <td>0.990006</td>\n",
       "      <td>0.367326</td>\n",
       "      <td>0.030594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_433385_split_00001</td>\n",
       "      <td>0.993435</td>\n",
       "      <td>0.106689</td>\n",
       "      <td>0.901929</td>\n",
       "      <td>0.093558</td>\n",
       "      <td>0.951990</td>\n",
       "      <td>0.427165</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_51681_split_00001</td>\n",
       "      <td>0.987489</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.912422</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.951743</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_91778_split_00001</td>\n",
       "      <td>0.996878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.978145</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.998216</td>\n",
       "      <td>0.364853</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   s  neg_ex_1  neg_field_1  neg_field_2  \\\n",
       "split                                                                      \n",
       "KNLK_58_175146_split_00001  0.997489  0.142857     0.954306     0.037660   \n",
       "KNLK_58_387596_split_00001  0.998368  0.000000     0.917805     0.092392   \n",
       "KNLK_58_433385_split_00001  0.993435  0.106689     0.901929     0.093558   \n",
       "KNLK_58_51681_split_00001   0.987489  0.000000     0.912422     0.000000   \n",
       "KNLK_58_91778_split_00001   0.996878  0.000000     0.978145     0.000000   \n",
       "\n",
       "                            neg_field_bf  pos_ex_1  pos_ex_2  \n",
       "split                                                         \n",
       "KNLK_58_175146_split_00001      0.974140  0.371077  0.189305  \n",
       "KNLK_58_387596_split_00001      0.990006  0.367326  0.030594  \n",
       "KNLK_58_433385_split_00001      0.951990  0.427165  0.000000  \n",
       "KNLK_58_51681_split_00001       0.951743  0.000000  0.000000  \n",
       "KNLK_58_91778_split_00001       0.998216  0.364853  0.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score_neg_ex_1</th>\n",
       "      <th>score_neg_field_1</th>\n",
       "      <th>score_neg_field_2</th>\n",
       "      <th>score_neg_field_bf</th>\n",
       "      <th>score_pos_ex_1</th>\n",
       "      <th>score_pos_ex_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_175146_split_00001</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_387596_split_00001</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_433385_split_00001</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_51681_split_00001</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_91778_split_00001</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "      <td>noncontam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           score_neg_ex_1 score_neg_field_1 score_neg_field_2  \\\n",
       "split                                                                           \n",
       "KNLK_58_175146_split_00001      noncontam         noncontam         noncontam   \n",
       "KNLK_58_387596_split_00001      noncontam         noncontam         noncontam   \n",
       "KNLK_58_433385_split_00001      noncontam         noncontam         noncontam   \n",
       "KNLK_58_51681_split_00001       noncontam         noncontam         noncontam   \n",
       "KNLK_58_91778_split_00001       noncontam         noncontam         noncontam   \n",
       "\n",
       "                           score_neg_field_bf score_pos_ex_1 score_pos_ex_2  \n",
       "split                                                                        \n",
       "KNLK_58_175146_split_00001          noncontam      noncontam      noncontam  \n",
       "KNLK_58_387596_split_00001          noncontam      noncontam      noncontam  \n",
       "KNLK_58_433385_split_00001          noncontam      noncontam      noncontam  \n",
       "KNLK_58_51681_split_00001           noncontam      noncontam      noncontam  \n",
       "KNLK_58_91778_split_00001           noncontam      noncontam      noncontam  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_91778_split_00001</td>\n",
       "      <td>Mycobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_51681_split_00001</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_175146_split_00001</td>\n",
       "      <td>T5virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_387596_split_00001</td>\n",
       "      <td>Unknown_Caudovirales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_433385_split_00001</td>\n",
       "      <td>Unknown_Verrucomicrobia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              genus\n",
       "split                                              \n",
       "KNLK_58_91778_split_00001             Mycobacterium\n",
       "KNLK_58_51681_split_00001                       NaN\n",
       "KNLK_58_175146_split_00001                  T5virus\n",
       "KNLK_58_387596_split_00001     Unknown_Caudovirales\n",
       "KNLK_58_433385_split_00001  Unknown_Verrucomicrobia"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# what contigs were reported by the detection method but not  by the coverage method\n",
    "detect_not_cov = list(set(detect_contam_contigs) - set(possible_contams_contigs))\n",
    "display(cov_means[cov_means.index.isin(detect_not_cov)].merge(tax, how='left', on='split'))\n",
    "display(cov_stds[cov_stds.index.isin(detect_not_cov)])\n",
    "display(detect[detect.index.isin(detect_not_cov)])\n",
    "display(iscontam[iscontam.index.isin(detect_not_cov)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_x</th>\n",
       "      <th>neg_ex_1_x</th>\n",
       "      <th>neg_field_1_x</th>\n",
       "      <th>neg_field_2_x</th>\n",
       "      <th>neg_field_bf_x</th>\n",
       "      <th>pos_ex_1_x</th>\n",
       "      <th>pos_ex_2_x</th>\n",
       "      <th>genus</th>\n",
       "      <th>s_y</th>\n",
       "      <th>neg_ex_1_y</th>\n",
       "      <th>neg_field_1_y</th>\n",
       "      <th>neg_field_2_y</th>\n",
       "      <th>neg_field_bf_y</th>\n",
       "      <th>pos_ex_1_y</th>\n",
       "      <th>pos_ex_2_y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_10037_split_00001</td>\n",
       "      <td>48.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.08</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Betaproteobacteria</td>\n",
       "      <td>79.15</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.03</td>\n",
       "      <td>1.40</td>\n",
       "      <td>0.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_10080_split_00001</td>\n",
       "      <td>1.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.66</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>1.26</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.96</td>\n",
       "      <td>0.00</td>\n",
       "      <td>13.14</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_10152_split_00001</td>\n",
       "      <td>14.14</td>\n",
       "      <td>4.88</td>\n",
       "      <td>51.37</td>\n",
       "      <td>0.13</td>\n",
       "      <td>88.73</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>6.53</td>\n",
       "      <td>6.01</td>\n",
       "      <td>26.03</td>\n",
       "      <td>1.60</td>\n",
       "      <td>40.38</td>\n",
       "      <td>2.16</td>\n",
       "      <td>0.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_105463_split_00001</td>\n",
       "      <td>2.30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.62</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.12</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Rickettsia</td>\n",
       "      <td>3.54</td>\n",
       "      <td>0.00</td>\n",
       "      <td>15.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>29.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_113171_split_00001</td>\n",
       "      <td>8.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.25</td>\n",
       "      <td>1.72</td>\n",
       "      <td>0.05</td>\n",
       "      <td>Nitrosomonas</td>\n",
       "      <td>5.81</td>\n",
       "      <td>0.37</td>\n",
       "      <td>4.74</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7.00</td>\n",
       "      <td>1.77</td>\n",
       "      <td>0.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_117233_split_00001</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7.67</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Betaproteobacteria</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_133487_split_00001</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.81</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>2.25</td>\n",
       "      <td>3.54</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.10</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_137858_split_00001</td>\n",
       "      <td>1.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.79</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.56</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_138203_split_00001</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Angomonas</td>\n",
       "      <td>1.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_138993_split_00001</td>\n",
       "      <td>3.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Caulobacter</td>\n",
       "      <td>1.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.53</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_1620_split_00001</td>\n",
       "      <td>12.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.12</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Methylobacterium</td>\n",
       "      <td>4.20</td>\n",
       "      <td>1.09</td>\n",
       "      <td>2.52</td>\n",
       "      <td>0.13</td>\n",
       "      <td>9.05</td>\n",
       "      <td>1.18</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_165019_split_00001</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>1.34</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_170698_split_00001</td>\n",
       "      <td>2.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.69</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.06</td>\n",
       "      <td>1.19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Methylobacterium</td>\n",
       "      <td>1.53</td>\n",
       "      <td>5.92</td>\n",
       "      <td>2.46</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.77</td>\n",
       "      <td>1.77</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_171115_split_00001</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.34</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_17269_split_00001</td>\n",
       "      <td>2.12</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.28</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>1.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.33</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_208916_split_00001</td>\n",
       "      <td>4.60</td>\n",
       "      <td>0.04</td>\n",
       "      <td>2.42</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>2.90</td>\n",
       "      <td>1.05</td>\n",
       "      <td>2.10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.24</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_210574_split_00001</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_211398_split_00001</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.99</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Sphingopyxis</td>\n",
       "      <td>2.46</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.54</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12.35</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_215264_split_00001</td>\n",
       "      <td>1.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>2.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.67</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_218090_split_00001</td>\n",
       "      <td>4.91</td>\n",
       "      <td>17.58</td>\n",
       "      <td>49.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>128.59</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Methylobacterium</td>\n",
       "      <td>2.30</td>\n",
       "      <td>13.23</td>\n",
       "      <td>26.62</td>\n",
       "      <td>0.00</td>\n",
       "      <td>55.88</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_227517_split_00001</td>\n",
       "      <td>11.29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>14.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Methylobacterium</td>\n",
       "      <td>5.13</td>\n",
       "      <td>2.56</td>\n",
       "      <td>3.77</td>\n",
       "      <td>0.29</td>\n",
       "      <td>10.61</td>\n",
       "      <td>1.07</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_232126_split_00001</td>\n",
       "      <td>3.22</td>\n",
       "      <td>2.45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.57</td>\n",
       "      <td>2.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.12</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_232350_split_00001</td>\n",
       "      <td>5.31</td>\n",
       "      <td>0.04</td>\n",
       "      <td>1.67</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.26</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>2.77</td>\n",
       "      <td>5.23</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.49</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_241716_split_00001</td>\n",
       "      <td>9.72</td>\n",
       "      <td>2.84</td>\n",
       "      <td>14.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>29.68</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bacteria</td>\n",
       "      <td>6.37</td>\n",
       "      <td>11.77</td>\n",
       "      <td>35.91</td>\n",
       "      <td>1.08</td>\n",
       "      <td>76.50</td>\n",
       "      <td>1.30</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_249037_split_00001</td>\n",
       "      <td>43.27</td>\n",
       "      <td>3.41</td>\n",
       "      <td>38.18</td>\n",
       "      <td>30.74</td>\n",
       "      <td>64.29</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bacteria</td>\n",
       "      <td>47.90</td>\n",
       "      <td>15.22</td>\n",
       "      <td>55.17</td>\n",
       "      <td>51.84</td>\n",
       "      <td>81.11</td>\n",
       "      <td>4.43</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_2572_split_00001</td>\n",
       "      <td>4.65</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>3.55</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.46</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_27344_split_00001</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_281399_split_00001</td>\n",
       "      <td>3.59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Proteobacteria</td>\n",
       "      <td>1.87</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.91</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_285526_split_00001</td>\n",
       "      <td>4.68</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bacteria</td>\n",
       "      <td>2.98</td>\n",
       "      <td>1.91</td>\n",
       "      <td>2.86</td>\n",
       "      <td>1.08</td>\n",
       "      <td>3.36</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_30996_split_00001</td>\n",
       "      <td>2.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>1.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_323115_split_00001</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.09</td>\n",
       "      <td>2.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>2.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.42</td>\n",
       "      <td>1.74</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_345941_split_00001</td>\n",
       "      <td>5.78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.87</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Erythrobacteraceae</td>\n",
       "      <td>3.25</td>\n",
       "      <td>1.16</td>\n",
       "      <td>5.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_347557_split_00001</td>\n",
       "      <td>1.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.92</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Bradyrhizobium</td>\n",
       "      <td>1.35</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.62</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_353100_split_00001</td>\n",
       "      <td>171.33</td>\n",
       "      <td>220.28</td>\n",
       "      <td>58.05</td>\n",
       "      <td>48.34</td>\n",
       "      <td>140.69</td>\n",
       "      <td>148.66</td>\n",
       "      <td>23.90</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>146.02</td>\n",
       "      <td>354.65</td>\n",
       "      <td>96.28</td>\n",
       "      <td>81.41</td>\n",
       "      <td>241.44</td>\n",
       "      <td>235.92</td>\n",
       "      <td>39.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_353100_split_00002</td>\n",
       "      <td>171.33</td>\n",
       "      <td>220.28</td>\n",
       "      <td>58.05</td>\n",
       "      <td>48.34</td>\n",
       "      <td>140.69</td>\n",
       "      <td>148.66</td>\n",
       "      <td>23.90</td>\n",
       "      <td>Lysobacter</td>\n",
       "      <td>146.02</td>\n",
       "      <td>354.65</td>\n",
       "      <td>96.28</td>\n",
       "      <td>81.41</td>\n",
       "      <td>241.44</td>\n",
       "      <td>235.92</td>\n",
       "      <td>39.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_35560_split_00001</td>\n",
       "      <td>19.12</td>\n",
       "      <td>0.07</td>\n",
       "      <td>3.89</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.65</td>\n",
       "      <td>0.45</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.90</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_362775_split_00001</td>\n",
       "      <td>3.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>29.35</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bradyrhizobiaceae</td>\n",
       "      <td>2.82</td>\n",
       "      <td>0.00</td>\n",
       "      <td>16.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_365020_split_00001</td>\n",
       "      <td>3.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.49</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>3.76</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>5.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_367774_split_00001</td>\n",
       "      <td>1.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Bradyrhizobium</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.75</td>\n",
       "      <td>2.54</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_367894_split_00001</td>\n",
       "      <td>4.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>2.76</td>\n",
       "      <td>1.29</td>\n",
       "      <td>1.71</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.58</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_377503_split_00001</td>\n",
       "      <td>6.09</td>\n",
       "      <td>636.63</td>\n",
       "      <td>159.87</td>\n",
       "      <td>150.20</td>\n",
       "      <td>445.02</td>\n",
       "      <td>423.65</td>\n",
       "      <td>64.82</td>\n",
       "      <td>Lysobacter</td>\n",
       "      <td>3.61</td>\n",
       "      <td>327.82</td>\n",
       "      <td>79.32</td>\n",
       "      <td>76.99</td>\n",
       "      <td>225.66</td>\n",
       "      <td>197.09</td>\n",
       "      <td>31.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_390948_split_00001</td>\n",
       "      <td>3.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7.52</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Betaproteobacteria</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.26</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_411033_split_00001</td>\n",
       "      <td>1.81</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bacteria</td>\n",
       "      <td>1.46</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.68</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_41674_split_00001</td>\n",
       "      <td>1.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>1.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_418965_split_00001</td>\n",
       "      <td>17.30</td>\n",
       "      <td>0.90</td>\n",
       "      <td>3.64</td>\n",
       "      <td>0.54</td>\n",
       "      <td>1.39</td>\n",
       "      <td>12.22</td>\n",
       "      <td>58.59</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>6.52</td>\n",
       "      <td>3.11</td>\n",
       "      <td>4.42</td>\n",
       "      <td>2.78</td>\n",
       "      <td>3.81</td>\n",
       "      <td>29.69</td>\n",
       "      <td>165.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_427171_split_00001</td>\n",
       "      <td>3.68</td>\n",
       "      <td>0.29</td>\n",
       "      <td>18.15</td>\n",
       "      <td>0.00</td>\n",
       "      <td>45.87</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Sphingomonadaceae</td>\n",
       "      <td>3.04</td>\n",
       "      <td>5.10</td>\n",
       "      <td>14.74</td>\n",
       "      <td>0.00</td>\n",
       "      <td>37.28</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_436901_split_00001</td>\n",
       "      <td>5.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12.66</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>3.04</td>\n",
       "      <td>3.08</td>\n",
       "      <td>9.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>21.44</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_452822_split_00001</td>\n",
       "      <td>11.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.71</td>\n",
       "      <td>0.59</td>\n",
       "      <td>5.42</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bacteria</td>\n",
       "      <td>5.56</td>\n",
       "      <td>1.74</td>\n",
       "      <td>5.98</td>\n",
       "      <td>2.65</td>\n",
       "      <td>12.72</td>\n",
       "      <td>0.92</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_464459_split_00001</td>\n",
       "      <td>27.44</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.57</td>\n",
       "      <td>1.62</td>\n",
       "      <td>3.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Sutterella</td>\n",
       "      <td>234.09</td>\n",
       "      <td>1.46</td>\n",
       "      <td>12.83</td>\n",
       "      <td>14.03</td>\n",
       "      <td>22.73</td>\n",
       "      <td>2.96</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_47619_split_00001</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_479137_split_00001</td>\n",
       "      <td>1.77</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.20</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Blastomonas</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.42</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_485138_split_00001</td>\n",
       "      <td>216.30</td>\n",
       "      <td>170.73</td>\n",
       "      <td>44.41</td>\n",
       "      <td>36.32</td>\n",
       "      <td>112.01</td>\n",
       "      <td>113.12</td>\n",
       "      <td>17.67</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>151.07</td>\n",
       "      <td>396.59</td>\n",
       "      <td>105.36</td>\n",
       "      <td>90.56</td>\n",
       "      <td>265.92</td>\n",
       "      <td>260.73</td>\n",
       "      <td>43.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_485138_split_00002</td>\n",
       "      <td>216.30</td>\n",
       "      <td>170.73</td>\n",
       "      <td>44.41</td>\n",
       "      <td>36.32</td>\n",
       "      <td>112.01</td>\n",
       "      <td>113.12</td>\n",
       "      <td>17.67</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>151.07</td>\n",
       "      <td>396.59</td>\n",
       "      <td>105.36</td>\n",
       "      <td>90.56</td>\n",
       "      <td>265.92</td>\n",
       "      <td>260.73</td>\n",
       "      <td>43.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_485138_split_00003</td>\n",
       "      <td>216.30</td>\n",
       "      <td>170.73</td>\n",
       "      <td>44.41</td>\n",
       "      <td>36.32</td>\n",
       "      <td>112.01</td>\n",
       "      <td>113.12</td>\n",
       "      <td>17.67</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>151.07</td>\n",
       "      <td>396.59</td>\n",
       "      <td>105.36</td>\n",
       "      <td>90.56</td>\n",
       "      <td>265.92</td>\n",
       "      <td>260.73</td>\n",
       "      <td>43.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_485138_split_00004</td>\n",
       "      <td>216.30</td>\n",
       "      <td>170.73</td>\n",
       "      <td>44.41</td>\n",
       "      <td>36.32</td>\n",
       "      <td>112.01</td>\n",
       "      <td>113.12</td>\n",
       "      <td>17.67</td>\n",
       "      <td>Aquabacterium</td>\n",
       "      <td>151.07</td>\n",
       "      <td>396.59</td>\n",
       "      <td>105.36</td>\n",
       "      <td>90.56</td>\n",
       "      <td>265.92</td>\n",
       "      <td>260.73</td>\n",
       "      <td>43.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_485138_split_00005</td>\n",
       "      <td>216.30</td>\n",
       "      <td>170.73</td>\n",
       "      <td>44.41</td>\n",
       "      <td>36.32</td>\n",
       "      <td>112.01</td>\n",
       "      <td>113.12</td>\n",
       "      <td>17.67</td>\n",
       "      <td>Lysobacter</td>\n",
       "      <td>151.07</td>\n",
       "      <td>396.59</td>\n",
       "      <td>105.36</td>\n",
       "      <td>90.56</td>\n",
       "      <td>265.92</td>\n",
       "      <td>260.73</td>\n",
       "      <td>43.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_485138_split_00006</td>\n",
       "      <td>216.30</td>\n",
       "      <td>170.73</td>\n",
       "      <td>44.41</td>\n",
       "      <td>36.32</td>\n",
       "      <td>112.01</td>\n",
       "      <td>113.12</td>\n",
       "      <td>17.67</td>\n",
       "      <td>Lysobacter</td>\n",
       "      <td>151.07</td>\n",
       "      <td>396.59</td>\n",
       "      <td>105.36</td>\n",
       "      <td>90.56</td>\n",
       "      <td>265.92</td>\n",
       "      <td>260.73</td>\n",
       "      <td>43.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_495785_split_00001</td>\n",
       "      <td>0.71</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.39</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_498868_split_00001</td>\n",
       "      <td>2.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Caulobacter</td>\n",
       "      <td>2.19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.62</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_505951_split_00001</td>\n",
       "      <td>7.81</td>\n",
       "      <td>553.89</td>\n",
       "      <td>152.78</td>\n",
       "      <td>131.76</td>\n",
       "      <td>399.32</td>\n",
       "      <td>397.41</td>\n",
       "      <td>58.62</td>\n",
       "      <td>Lysobacter</td>\n",
       "      <td>5.01</td>\n",
       "      <td>434.84</td>\n",
       "      <td>119.57</td>\n",
       "      <td>101.08</td>\n",
       "      <td>318.23</td>\n",
       "      <td>304.56</td>\n",
       "      <td>47.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_506749_split_00001</td>\n",
       "      <td>2.18</td>\n",
       "      <td>0.00</td>\n",
       "      <td>114.91</td>\n",
       "      <td>0.47</td>\n",
       "      <td>2.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Sporanaerobacter</td>\n",
       "      <td>2.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>276.58</td>\n",
       "      <td>2.65</td>\n",
       "      <td>6.67</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_511747_split_00001</td>\n",
       "      <td>2.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.89</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.53</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "      <td>0.96</td>\n",
       "      <td>0.65</td>\n",
       "      <td>4.38</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7.21</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_535926_split_00001</td>\n",
       "      <td>2.49</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Bradyrhizobium</td>\n",
       "      <td>1.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_57490_split_00001</td>\n",
       "      <td>2.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Unknown_Bacteria</td>\n",
       "      <td>1.90</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.92</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.25</td>\n",
       "      <td>1.71</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_59495_split_00001</td>\n",
       "      <td>2.71</td>\n",
       "      <td>6.53</td>\n",
       "      <td>16.23</td>\n",
       "      <td>1.53</td>\n",
       "      <td>33.95</td>\n",
       "      <td>2.04</td>\n",
       "      <td>0.05</td>\n",
       "      <td>Sphingomonas</td>\n",
       "      <td>2.10</td>\n",
       "      <td>7.44</td>\n",
       "      <td>10.80</td>\n",
       "      <td>1.45</td>\n",
       "      <td>25.64</td>\n",
       "      <td>4.16</td>\n",
       "      <td>0.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_68143_split_00001</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_89437_split_00001</td>\n",
       "      <td>1.72</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.71</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>1.05</td>\n",
       "      <td>2.67</td>\n",
       "      <td>1.61</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.51</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_95132_split_00001</td>\n",
       "      <td>3.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.61</td>\n",
       "      <td>2.51</td>\n",
       "      <td>3.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.70</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               s_x  neg_ex_1_x  neg_field_1_x  neg_field_2_x  \\\n",
       "split                                                                          \n",
       "KNLK_58_10037_split_00001    48.84        0.00           0.02           0.00   \n",
       "KNLK_58_10080_split_00001     1.97        0.00           1.57           0.00   \n",
       "KNLK_58_10152_split_00001    14.14        4.88          51.37           0.13   \n",
       "KNLK_58_105463_split_00001    2.30        0.00           1.62           0.00   \n",
       "KNLK_58_113171_split_00001    8.95        0.00           4.97           0.00   \n",
       "KNLK_58_117233_split_00001    2.20        0.00           3.93           0.00   \n",
       "KNLK_58_133487_split_00001    1.22        0.17           0.48           0.00   \n",
       "KNLK_58_137858_split_00001    1.21        0.00           0.95           0.00   \n",
       "KNLK_58_138203_split_00001    0.00        0.00           0.00           0.00   \n",
       "KNLK_58_138993_split_00001    3.18        0.00           0.27           0.00   \n",
       "KNLK_58_1620_split_00001     12.20        0.00           1.12           0.00   \n",
       "KNLK_58_165019_split_00001    1.76        0.00           0.00           0.00   \n",
       "KNLK_58_170698_split_00001    2.32        0.00           0.69           0.00   \n",
       "KNLK_58_171115_split_00001    0.00        0.00           0.00           0.00   \n",
       "KNLK_58_17269_split_00001     2.12        0.00           0.00           0.00   \n",
       "KNLK_58_208916_split_00001    4.60        0.04           2.42           0.00   \n",
       "KNLK_58_210574_split_00001    0.00        0.00           0.00           0.00   \n",
       "KNLK_58_211398_split_00001    2.11        0.00           1.99           0.00   \n",
       "KNLK_58_215264_split_00001    1.98        0.00           1.13           0.00   \n",
       "KNLK_58_218090_split_00001    4.91       17.58          49.47           0.00   \n",
       "KNLK_58_227517_split_00001   11.29        0.00           3.18           0.00   \n",
       "KNLK_58_232126_split_00001    3.22        2.45           0.00           0.00   \n",
       "KNLK_58_232350_split_00001    5.31        0.04           1.67           0.00   \n",
       "KNLK_58_241716_split_00001    9.72        2.84          14.06           0.00   \n",
       "KNLK_58_249037_split_00001   43.27        3.41          38.18          30.74   \n",
       "KNLK_58_2572_split_00001      4.65        0.00           1.09           0.00   \n",
       "KNLK_58_27344_split_00001     1.76        0.00           0.58           0.00   \n",
       "KNLK_58_281399_split_00001    3.59        0.00           3.32           0.00   \n",
       "KNLK_58_285526_split_00001    4.68        0.00           2.36           0.00   \n",
       "KNLK_58_30996_split_00001     2.21        0.00           0.00           0.00   \n",
       "KNLK_58_323115_split_00001    2.66        0.00           0.00           0.00   \n",
       "KNLK_58_345941_split_00001    5.78        0.00           2.87           0.00   \n",
       "KNLK_58_347557_split_00001    1.86        0.00           1.92           0.00   \n",
       "KNLK_58_353100_split_00001  171.33      220.28          58.05          48.34   \n",
       "KNLK_58_353100_split_00002  171.33      220.28          58.05          48.34   \n",
       "KNLK_58_35560_split_00001    19.12        0.07           3.89           0.00   \n",
       "KNLK_58_362775_split_00001    3.06        0.00          29.35           0.00   \n",
       "KNLK_58_365020_split_00001    3.57        0.00           0.00           0.00   \n",
       "KNLK_58_367774_split_00001    1.80        0.00           2.84           0.00   \n",
       "KNLK_58_367894_split_00001    4.51        0.00           0.66           0.00   \n",
       "KNLK_58_377503_split_00001    6.09      636.63         159.87         150.20   \n",
       "KNLK_58_390948_split_00001    3.03        0.00           7.52           0.00   \n",
       "KNLK_58_411033_split_00001    1.81        0.00           0.72           0.00   \n",
       "KNLK_58_41674_split_00001     1.23        0.00           0.53           0.01   \n",
       "KNLK_58_418965_split_00001   17.30        0.90           3.64           0.54   \n",
       "KNLK_58_427171_split_00001    3.68        0.29          18.15           0.00   \n",
       "KNLK_58_436901_split_00001    5.01        0.00           4.85           0.00   \n",
       "KNLK_58_452822_split_00001   11.98        0.00           4.71           0.59   \n",
       "KNLK_58_464459_split_00001   27.44        0.00           2.57           1.62   \n",
       "KNLK_58_47619_split_00001     0.00        0.00           0.00           0.00   \n",
       "KNLK_58_479137_split_00001    1.77        0.00           0.00           0.00   \n",
       "KNLK_58_485138_split_00001  216.30      170.73          44.41          36.32   \n",
       "KNLK_58_485138_split_00002  216.30      170.73          44.41          36.32   \n",
       "KNLK_58_485138_split_00003  216.30      170.73          44.41          36.32   \n",
       "KNLK_58_485138_split_00004  216.30      170.73          44.41          36.32   \n",
       "KNLK_58_485138_split_00005  216.30      170.73          44.41          36.32   \n",
       "KNLK_58_485138_split_00006  216.30      170.73          44.41          36.32   \n",
       "KNLK_58_495785_split_00001    0.71        0.00           0.00           0.00   \n",
       "KNLK_58_498868_split_00001    2.21        0.00           2.25           0.00   \n",
       "KNLK_58_505951_split_00001    7.81      553.89         152.78         131.76   \n",
       "KNLK_58_506749_split_00001    2.18        0.00         114.91           0.47   \n",
       "KNLK_58_511747_split_00001    2.86        0.00           3.89           0.00   \n",
       "KNLK_58_535926_split_00001    2.49        0.00           0.00           0.68   \n",
       "KNLK_58_57490_split_00001     2.27        0.00           1.76           0.00   \n",
       "KNLK_58_59495_split_00001     2.71        6.53          16.23           1.53   \n",
       "KNLK_58_68143_split_00001     0.42        0.00           0.00           0.00   \n",
       "KNLK_58_89437_split_00001     1.72        0.00           0.00           0.00   \n",
       "KNLK_58_95132_split_00001     3.47        0.00           1.85           0.00   \n",
       "\n",
       "                            neg_field_bf_x  pos_ex_1_x  pos_ex_2_x  \\\n",
       "split                                                                \n",
       "KNLK_58_10037_split_00001             2.08        0.41        0.00   \n",
       "KNLK_58_10080_split_00001             3.66        0.01        0.01   \n",
       "KNLK_58_10152_split_00001            88.73        2.20        0.00   \n",
       "KNLK_58_105463_split_00001            2.12        0.00        0.00   \n",
       "KNLK_58_113171_split_00001            9.25        1.72        0.05   \n",
       "KNLK_58_117233_split_00001            7.67        0.00        0.00   \n",
       "KNLK_58_133487_split_00001            1.81        0.14        0.00   \n",
       "KNLK_58_137858_split_00001            4.79        0.00        0.00   \n",
       "KNLK_58_138203_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_138993_split_00001            2.21        0.00        0.00   \n",
       "KNLK_58_1620_split_00001              6.98        0.00        0.00   \n",
       "KNLK_58_165019_split_00001            1.60        0.00        0.00   \n",
       "KNLK_58_170698_split_00001            1.06        1.19        0.00   \n",
       "KNLK_58_171115_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_17269_split_00001             0.00        1.28        0.00   \n",
       "KNLK_58_208916_split_00001            8.86        0.00        0.00   \n",
       "KNLK_58_210574_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_211398_split_00001            1.47        0.00        0.00   \n",
       "KNLK_58_215264_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_218090_split_00001          128.59        0.24        0.00   \n",
       "KNLK_58_227517_split_00001           14.76        0.00        0.00   \n",
       "KNLK_58_232126_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_232350_split_00001            3.26        0.02        0.00   \n",
       "KNLK_58_241716_split_00001           29.68        0.00        0.00   \n",
       "KNLK_58_249037_split_00001           64.29        0.83        0.00   \n",
       "KNLK_58_2572_split_00001              3.84        0.00        0.00   \n",
       "KNLK_58_27344_split_00001             0.00        0.00        0.00   \n",
       "KNLK_58_281399_split_00001            6.97        0.00        0.00   \n",
       "KNLK_58_285526_split_00001            0.18        0.00        0.00   \n",
       "KNLK_58_30996_split_00001             3.08        0.00        0.00   \n",
       "KNLK_58_323115_split_00001            1.09        2.75        0.00   \n",
       "KNLK_58_345941_split_00001            3.25        0.00        0.00   \n",
       "KNLK_58_347557_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_353100_split_00001          140.69      148.66       23.90   \n",
       "KNLK_58_353100_split_00002          140.69      148.66       23.90   \n",
       "KNLK_58_35560_split_00001             2.95        0.00        0.00   \n",
       "KNLK_58_362775_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_365020_split_00001            3.49        0.00        0.00   \n",
       "KNLK_58_367774_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_367894_split_00001            2.31        0.00        0.00   \n",
       "KNLK_58_377503_split_00001          445.02      423.65       64.82   \n",
       "KNLK_58_390948_split_00001            3.25        0.00        0.00   \n",
       "KNLK_58_411033_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_41674_split_00001             0.00        0.00        0.00   \n",
       "KNLK_58_418965_split_00001            1.39       12.22       58.59   \n",
       "KNLK_58_427171_split_00001           45.87        0.00        0.00   \n",
       "KNLK_58_436901_split_00001           12.66        0.00        0.00   \n",
       "KNLK_58_452822_split_00001            5.42        0.00        0.00   \n",
       "KNLK_58_464459_split_00001            3.75        0.00        0.00   \n",
       "KNLK_58_47619_split_00001             0.00        0.00        0.00   \n",
       "KNLK_58_479137_split_00001            2.20        0.00        0.00   \n",
       "KNLK_58_485138_split_00001          112.01      113.12       17.67   \n",
       "KNLK_58_485138_split_00002          112.01      113.12       17.67   \n",
       "KNLK_58_485138_split_00003          112.01      113.12       17.67   \n",
       "KNLK_58_485138_split_00004          112.01      113.12       17.67   \n",
       "KNLK_58_485138_split_00005          112.01      113.12       17.67   \n",
       "KNLK_58_485138_split_00006          112.01      113.12       17.67   \n",
       "KNLK_58_495785_split_00001            0.51        0.00        0.00   \n",
       "KNLK_58_498868_split_00001            3.13        0.00        0.00   \n",
       "KNLK_58_505951_split_00001          399.32      397.41       58.62   \n",
       "KNLK_58_506749_split_00001            2.04        0.00        0.00   \n",
       "KNLK_58_511747_split_00001            9.53        0.00        0.00   \n",
       "KNLK_58_535926_split_00001            0.00        0.00        0.00   \n",
       "KNLK_58_57490_split_00001             0.17        0.42        0.00   \n",
       "KNLK_58_59495_split_00001            33.95        2.04        0.05   \n",
       "KNLK_58_68143_split_00001             0.30        0.00        0.00   \n",
       "KNLK_58_89437_split_00001             1.71        0.00        0.00   \n",
       "KNLK_58_95132_split_00001             0.00        0.00        0.00   \n",
       "\n",
       "                                                  genus     s_y  neg_ex_1_y  \\\n",
       "split                                                                         \n",
       "KNLK_58_10037_split_00001    Unknown_Betaproteobacteria   79.15        0.00   \n",
       "KNLK_58_10080_split_00001   Unknown_Alphaproteobacteria    1.26        0.00   \n",
       "KNLK_58_10152_split_00001   Unknown_Alphaproteobacteria    6.53        6.01   \n",
       "KNLK_58_105463_split_00001                   Rickettsia    3.54        0.00   \n",
       "KNLK_58_113171_split_00001                 Nitrosomonas    5.81        0.37   \n",
       "KNLK_58_117233_split_00001   Unknown_Betaproteobacteria    0.93        0.00   \n",
       "KNLK_58_133487_split_00001                Mycobacterium    2.25        3.54   \n",
       "KNLK_58_137858_split_00001                Aquabacterium    0.74        0.00   \n",
       "KNLK_58_138203_split_00001                    Angomonas    1.18        0.00   \n",
       "KNLK_58_138993_split_00001                  Caulobacter    1.93        0.00   \n",
       "KNLK_58_1620_split_00001               Methylobacterium    4.20        1.09   \n",
       "KNLK_58_165019_split_00001                Mycobacterium    1.34        0.00   \n",
       "KNLK_58_170698_split_00001             Methylobacterium    1.53        5.92   \n",
       "KNLK_58_171115_split_00001                          NaN    1.34        0.00   \n",
       "KNLK_58_17269_split_00001                 Mycobacterium    1.40        0.00   \n",
       "KNLK_58_208916_split_00001  Unknown_Alphaproteobacteria    2.90        1.05   \n",
       "KNLK_58_210574_split_00001                          NaN    0.36        0.00   \n",
       "KNLK_58_211398_split_00001                 Sphingopyxis    2.46        0.00   \n",
       "KNLK_58_215264_split_00001                Mycobacterium    2.13        0.00   \n",
       "KNLK_58_218090_split_00001             Methylobacterium    2.30       13.23   \n",
       "KNLK_58_227517_split_00001             Methylobacterium    5.13        2.56   \n",
       "KNLK_58_232126_split_00001                          NaN    2.57        2.75   \n",
       "KNLK_58_232350_split_00001  Unknown_Alphaproteobacteria    2.77        5.23   \n",
       "KNLK_58_241716_split_00001             Unknown_Bacteria    6.37       11.77   \n",
       "KNLK_58_249037_split_00001             Unknown_Bacteria   47.90       15.22   \n",
       "KNLK_58_2572_split_00001                  Aquabacterium    3.55        0.00   \n",
       "KNLK_58_27344_split_00001                 Mycobacterium    1.32        0.00   \n",
       "KNLK_58_281399_split_00001       Unknown_Proteobacteria    1.87        0.00   \n",
       "KNLK_58_285526_split_00001             Unknown_Bacteria    2.98        1.91   \n",
       "KNLK_58_30996_split_00001                 Mycobacterium    1.06        0.00   \n",
       "KNLK_58_323115_split_00001                Mycobacterium    2.36        0.00   \n",
       "KNLK_58_345941_split_00001   Unknown_Erythrobacteraceae    3.25        1.16   \n",
       "KNLK_58_347557_split_00001               Bradyrhizobium    1.35        0.00   \n",
       "KNLK_58_353100_split_00001                Aquabacterium  146.02      354.65   \n",
       "KNLK_58_353100_split_00002                   Lysobacter  146.02      354.65   \n",
       "KNLK_58_35560_split_00001                           NaN   12.65        0.45   \n",
       "KNLK_58_362775_split_00001    Unknown_Bradyrhizobiaceae    2.82        0.00   \n",
       "KNLK_58_365020_split_00001                Mycobacterium    3.76        0.52   \n",
       "KNLK_58_367774_split_00001               Bradyrhizobium    1.60        0.75   \n",
       "KNLK_58_367894_split_00001                Mycobacterium    2.76        1.29   \n",
       "KNLK_58_377503_split_00001                   Lysobacter    3.61      327.82   \n",
       "KNLK_58_390948_split_00001   Unknown_Betaproteobacteria    1.32        0.00   \n",
       "KNLK_58_411033_split_00001             Unknown_Bacteria    1.46        0.00   \n",
       "KNLK_58_41674_split_00001   Unknown_Alphaproteobacteria    1.09        0.00   \n",
       "KNLK_58_418965_split_00001                  Pseudomonas    6.52        3.11   \n",
       "KNLK_58_427171_split_00001    Unknown_Sphingomonadaceae    3.04        5.10   \n",
       "KNLK_58_436901_split_00001  Unknown_Alphaproteobacteria    3.04        3.08   \n",
       "KNLK_58_452822_split_00001             Unknown_Bacteria    5.56        1.74   \n",
       "KNLK_58_464459_split_00001                   Sutterella  234.09        1.46   \n",
       "KNLK_58_47619_split_00001                 Mycobacterium    0.43        0.00   \n",
       "KNLK_58_479137_split_00001                  Blastomonas    1.32        0.00   \n",
       "KNLK_58_485138_split_00001                Aquabacterium  151.07      396.59   \n",
       "KNLK_58_485138_split_00002                Aquabacterium  151.07      396.59   \n",
       "KNLK_58_485138_split_00003                Aquabacterium  151.07      396.59   \n",
       "KNLK_58_485138_split_00004                Aquabacterium  151.07      396.59   \n",
       "KNLK_58_485138_split_00005                   Lysobacter  151.07      396.59   \n",
       "KNLK_58_485138_split_00006                   Lysobacter  151.07      396.59   \n",
       "KNLK_58_495785_split_00001                Mycobacterium    1.32        0.00   \n",
       "KNLK_58_498868_split_00001                  Caulobacter    2.19        0.00   \n",
       "KNLK_58_505951_split_00001                   Lysobacter    5.01      434.84   \n",
       "KNLK_58_506749_split_00001             Sporanaerobacter    2.06        0.00   \n",
       "KNLK_58_511747_split_00001  Unknown_Alphaproteobacteria    0.96        0.65   \n",
       "KNLK_58_535926_split_00001               Bradyrhizobium    1.86        0.00   \n",
       "KNLK_58_57490_split_00001              Unknown_Bacteria    1.90        0.00   \n",
       "KNLK_58_59495_split_00001                  Sphingomonas    2.10        7.44   \n",
       "KNLK_58_68143_split_00001                 Mycobacterium    0.93        0.00   \n",
       "KNLK_58_89437_split_00001                 Mycobacterium    1.05        2.67   \n",
       "KNLK_58_95132_split_00001                           NaN    2.61        2.51   \n",
       "\n",
       "                            neg_field_1_y  neg_field_2_y  neg_field_bf_y  \\\n",
       "split                                                                      \n",
       "KNLK_58_10037_split_00001            0.75           0.00            3.03   \n",
       "KNLK_58_10080_split_00001            4.96           0.00           13.14   \n",
       "KNLK_58_10152_split_00001           26.03           1.60           40.38   \n",
       "KNLK_58_105463_split_00001          15.27           0.00           29.86   \n",
       "KNLK_58_113171_split_00001           4.74           0.00            7.00   \n",
       "KNLK_58_117233_split_00001           3.33           0.00            8.18   \n",
       "KNLK_58_133487_split_00001           1.22           0.00            4.10   \n",
       "KNLK_58_137858_split_00001           1.56           0.00            3.80   \n",
       "KNLK_58_138203_split_00001           0.00           0.00            0.00   \n",
       "KNLK_58_138993_split_00001           1.45           0.00            4.53   \n",
       "KNLK_58_1620_split_00001             2.52           0.13            9.05   \n",
       "KNLK_58_165019_split_00001           0.32           0.00            3.51   \n",
       "KNLK_58_170698_split_00001           2.46           0.00            3.77   \n",
       "KNLK_58_171115_split_00001           0.00           0.00            0.00   \n",
       "KNLK_58_17269_split_00001            0.00           0.00            0.00   \n",
       "KNLK_58_208916_split_00001           2.10           0.00            8.24   \n",
       "KNLK_58_210574_split_00001           0.00           0.00            0.00   \n",
       "KNLK_58_211398_split_00001           8.54           0.00           12.35   \n",
       "KNLK_58_215264_split_00001           1.67           0.00            0.70   \n",
       "KNLK_58_218090_split_00001          26.62           0.00           55.88   \n",
       "KNLK_58_227517_split_00001           3.77           0.29           10.61   \n",
       "KNLK_58_232126_split_00001           0.00           1.27            0.00   \n",
       "KNLK_58_232350_split_00001           2.11           0.00            6.49   \n",
       "KNLK_58_241716_split_00001          35.91           1.08           76.50   \n",
       "KNLK_58_249037_split_00001          55.17          51.84           81.11   \n",
       "KNLK_58_2572_split_00001             1.59           0.00            3.46   \n",
       "KNLK_58_27344_split_00001            2.08           0.00            1.06   \n",
       "KNLK_58_281399_split_00001           4.04           0.00            4.91   \n",
       "KNLK_58_285526_split_00001           2.86           1.08            3.36   \n",
       "KNLK_58_30996_split_00001            0.60           0.00            3.84   \n",
       "KNLK_58_323115_split_00001           1.31           0.00            1.42   \n",
       "KNLK_58_345941_split_00001           5.06           0.00            6.01   \n",
       "KNLK_58_347557_split_00001           1.62           0.00            0.00   \n",
       "KNLK_58_353100_split_00001          96.28          81.41          241.44   \n",
       "KNLK_58_353100_split_00002          96.28          81.41          241.44   \n",
       "KNLK_58_35560_split_00001            2.20           0.00            4.90   \n",
       "KNLK_58_362775_split_00001          16.84           0.00            0.00   \n",
       "KNLK_58_365020_split_00001           0.00           0.00            5.75   \n",
       "KNLK_58_367774_split_00001           2.54           0.00            0.57   \n",
       "KNLK_58_367894_split_00001           1.71           0.00            6.58   \n",
       "KNLK_58_377503_split_00001          79.32          76.99          225.66   \n",
       "KNLK_58_390948_split_00001           4.40           0.00            4.26   \n",
       "KNLK_58_411033_split_00001           1.68           0.52            0.41   \n",
       "KNLK_58_41674_split_00001            0.78           0.52            0.00   \n",
       "KNLK_58_418965_split_00001           4.42           2.78            3.81   \n",
       "KNLK_58_427171_split_00001          14.74           0.00           37.28   \n",
       "KNLK_58_436901_split_00001           9.18           0.00           21.44   \n",
       "KNLK_58_452822_split_00001           5.98           2.65           12.72   \n",
       "KNLK_58_464459_split_00001          12.83          14.03           22.73   \n",
       "KNLK_58_47619_split_00001            0.00           0.00            0.00   \n",
       "KNLK_58_479137_split_00001           0.32           0.00            3.42   \n",
       "KNLK_58_485138_split_00001         105.36          90.56          265.92   \n",
       "KNLK_58_485138_split_00002         105.36          90.56          265.92   \n",
       "KNLK_58_485138_split_00003         105.36          90.56          265.92   \n",
       "KNLK_58_485138_split_00004         105.36          90.56          265.92   \n",
       "KNLK_58_485138_split_00005         105.36          90.56          265.92   \n",
       "KNLK_58_485138_split_00006         105.36          90.56          265.92   \n",
       "KNLK_58_495785_split_00001           0.00           0.00            2.39   \n",
       "KNLK_58_498868_split_00001           1.62           0.00            4.51   \n",
       "KNLK_58_505951_split_00001         119.57         101.08          318.23   \n",
       "KNLK_58_506749_split_00001         276.58           2.65            6.67   \n",
       "KNLK_58_511747_split_00001           4.38           0.00            7.21   \n",
       "KNLK_58_535926_split_00001           0.00           0.84            0.00   \n",
       "KNLK_58_57490_split_00001            1.92           0.00            1.25   \n",
       "KNLK_58_59495_split_00001           10.80           1.45           25.64   \n",
       "KNLK_58_68143_split_00001            0.00           0.00            2.29   \n",
       "KNLK_58_89437_split_00001            1.61           0.00            2.51   \n",
       "KNLK_58_95132_split_00001            3.95           0.00            1.70   \n",
       "\n",
       "                            pos_ex_1_y  pos_ex_2_y  \n",
       "split                                               \n",
       "KNLK_58_10037_split_00001         1.40        0.65  \n",
       "KNLK_58_10080_split_00001         0.44        0.44  \n",
       "KNLK_58_10152_split_00001         2.16        0.64  \n",
       "KNLK_58_105463_split_00001        0.00        0.00  \n",
       "KNLK_58_113171_split_00001        1.77        0.56  \n",
       "KNLK_58_117233_split_00001        0.00        0.00  \n",
       "KNLK_58_133487_split_00001        0.80        0.00  \n",
       "KNLK_58_137858_split_00001        0.00        0.00  \n",
       "KNLK_58_138203_split_00001        0.00        0.00  \n",
       "KNLK_58_138993_split_00001        0.00        0.00  \n",
       "KNLK_58_1620_split_00001          1.18        0.00  \n",
       "KNLK_58_165019_split_00001        0.00        0.00  \n",
       "KNLK_58_170698_split_00001        1.77        0.00  \n",
       "KNLK_58_171115_split_00001        0.00        0.00  \n",
       "KNLK_58_17269_split_00001         2.33        0.00  \n",
       "KNLK_58_208916_split_00001        0.00        0.00  \n",
       "KNLK_58_210574_split_00001        0.00        0.00  \n",
       "KNLK_58_211398_split_00001        0.00        0.00  \n",
       "KNLK_58_215264_split_00001        0.00        0.00  \n",
       "KNLK_58_218090_split_00001        0.89        0.81  \n",
       "KNLK_58_227517_split_00001        1.07        0.29  \n",
       "KNLK_58_232126_split_00001        1.12        0.00  \n",
       "KNLK_58_232350_split_00001        1.30        0.40  \n",
       "KNLK_58_241716_split_00001        1.30        0.29  \n",
       "KNLK_58_249037_split_00001        4.43        0.00  \n",
       "KNLK_58_2572_split_00001          0.00        0.00  \n",
       "KNLK_58_27344_split_00001         0.00        0.00  \n",
       "KNLK_58_281399_split_00001        0.33        0.00  \n",
       "KNLK_58_285526_split_00001        0.90        0.00  \n",
       "KNLK_58_30996_split_00001         0.00        0.00  \n",
       "KNLK_58_323115_split_00001        1.74        0.00  \n",
       "KNLK_58_345941_split_00001        0.00        0.00  \n",
       "KNLK_58_347557_split_00001        0.00        0.00  \n",
       "KNLK_58_353100_split_00001      235.92       39.65  \n",
       "KNLK_58_353100_split_00002      235.92       39.65  \n",
       "KNLK_58_35560_split_00001         0.58        0.00  \n",
       "KNLK_58_362775_split_00001        0.00        0.00  \n",
       "KNLK_58_365020_split_00001        0.00        0.00  \n",
       "KNLK_58_367774_split_00001        0.00        0.00  \n",
       "KNLK_58_367894_split_00001        0.00        0.00  \n",
       "KNLK_58_377503_split_00001      197.09       31.90  \n",
       "KNLK_58_390948_split_00001        0.00        0.00  \n",
       "KNLK_58_411033_split_00001        0.00        0.00  \n",
       "KNLK_58_41674_split_00001         0.00        0.00  \n",
       "KNLK_58_418965_split_00001       29.69      165.78  \n",
       "KNLK_58_427171_split_00001        0.42        0.37  \n",
       "KNLK_58_436901_split_00001        0.64        0.28  \n",
       "KNLK_58_452822_split_00001        0.92        0.00  \n",
       "KNLK_58_464459_split_00001        2.96        0.63  \n",
       "KNLK_58_47619_split_00001         0.00        0.00  \n",
       "KNLK_58_479137_split_00001        0.00        0.00  \n",
       "KNLK_58_485138_split_00001      260.73       43.79  \n",
       "KNLK_58_485138_split_00002      260.73       43.79  \n",
       "KNLK_58_485138_split_00003      260.73       43.79  \n",
       "KNLK_58_485138_split_00004      260.73       43.79  \n",
       "KNLK_58_485138_split_00005      260.73       43.79  \n",
       "KNLK_58_485138_split_00006      260.73       43.79  \n",
       "KNLK_58_495785_split_00001        0.00        0.00  \n",
       "KNLK_58_498868_split_00001        0.00        0.00  \n",
       "KNLK_58_505951_split_00001      304.56       47.07  \n",
       "KNLK_58_506749_split_00001        0.00        0.00  \n",
       "KNLK_58_511747_split_00001        0.00        0.00  \n",
       "KNLK_58_535926_split_00001        0.00        0.00  \n",
       "KNLK_58_57490_split_00001         1.71        0.00  \n",
       "KNLK_58_59495_split_00001         4.16        0.45  \n",
       "KNLK_58_68143_split_00001         0.00        0.00  \n",
       "KNLK_58_89437_split_00001         0.00        0.00  \n",
       "KNLK_58_95132_split_00001         0.47        0.00  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "1.6099963682585798"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "0.8522733905328898"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# which contigs were reported as possible contams by the coverage method but not by detection? (suggesting they may not actually be contaminants)\n",
    "display(cov_means[cov_means.index.isin(not_detect)].merge(tax, how='left', on='split').merge(cov_stds, how='left', on='split').round(2))\n",
    "#display(possible_contams[possible_contams.index.isin(not_detect)])\n",
    "display(354.65/220.28)\n",
    "display(146.02/171.33)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "collection_nodetect = pd.DataFrame(not_detect, columns=['split'])\n",
    "collection_nodetect['bin'] = 'not_detect'\n",
    "collection_nodetect_file = '/Users/rosekantor/data/awtp2_metagenomics/test_decontamination/contam_no_detect_collection.txt'\n",
    "collection_nodetect.to_csv(collection_nodetect_file, header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_10152_split_00001</td>\n",
       "      <td>Unknown_Alphaproteobacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 genus\n",
       "split                                                 \n",
       "KNLK_58_10152_split_00001  Unknown_Alphaproteobacteria"
      ]
     },
     "execution_count": 314,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# special case: KNLK_58_10152_split_00001 has med-high pretty even coverage in samples and some controls.\n",
    "# Coverage ends right on either side of the genes, doesn't extend to the edges of the contig\n",
    "# detection is 1 in sample and next highest control is 0.87 (so it didn't pass the \"detection\" threshold)\n",
    "# in Anvio, this contig clusters in a bin that I wasn't sure about- maybe contaminant, maybe not\n",
    "# This contig has a transposase! It is possibly not a contaminant then... just similar to something in contaminants?\n",
    "not_detect_df[not_detect_df.index == 'KNLK_58_10152_split_00001']\n",
    "tax[tax.index == 'KNLK_58_10152_split_00001']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x148422810>"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAD4CAYAAADFAawfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAQBElEQVR4nO3df4zkd13H8eeblqZHtx6VwoRckYUI1coKeBP8QSKzrZDKEYhItE0h1gCbGCnVHMYjxqAmxP5TDYn84YkEwo9O4KAGOQWJdCQYQHdpy7Y9SqAc0KteAeFgmgtw5O0fMwe3d9/d73fLfGc+O/d8JJObmc9n9/ve93z2ldnvfL/3jcxEklSux8y6AEnS1gxqSSqcQS1JhTOoJalwBrUkFe7CNr7p5ZdfnouLixuee+SRR7jkkkva2NyOZl+q2Zdz2ZNq89KXtbW1b2TmE6vGWgnqxcVFVldXNzw3GAzo9XptbG5Hsy/V7Mu57Em1eelLRHxlszF3fUhS4QxqSSqcQS1JhTOoJalwBrUkFc6glqTCNQrqiPjjiLg3Iu6JiNsi4uK2C5MkjdQGdUTsAV4PdDPzWcAFwHVtFyZJGmm66+NCYFdEXAg8DniovZIkSWeKJhcOiIibgTcDJ4F/y8wbKuasACsAnU5nb7/f3zA+HA5ZWFio3db6sRONCp+0pT27Z7Ldpn0539iXc9mTavPSl+Xl5bXM7FaN1QZ1RFwGfAD4XeDbwPuBQ5n57s2+ptvt5qM9hXzxwOHaOW04esu+mWx3Xk5/nTT7ci57Um1e+hIRmwZ1k10fvwF8OTO/npk/AD4I/NokC5Qkba5JUH8V+JWIeFxEBHANcKTdsiRJp9UGdWZ+BjgEfBZYH3/NwZbrkiSNNfpvTjPzTcCbWq5FklTBMxMlqXAGtSQVzqCWpMIZ1JJUOINakgpnUEtS4QxqSSqcQS1JhTOoJalwBrUkFc6glqTCGdSSVDiDWpIKZ1BLUuEMakkqnEEtSYWrDeqIuDIi7jrj9p2I+KNpFCdJanCFl8y8H3gOQERcABwDbm+5LknS2HZ3fVwDfCkzv9JGMZKkc203qK8DbmujEElStcjMZhMjLgIeAn4hM49XjK8AKwCdTmdvv9/fMD4cDllYWKjdzvqxE43qmbSlPbtnst2mfTnf2Jdz2ZNq89KX5eXltczsVo1tJ6hfBvxhZr6obm63283V1dUNzw0GA3q9Xu12Fg8cblTPpB29Zd9Mttu0L+cb+3Iue1JtXvoSEZsG9XZ2fVyPuz0kaeoaBXVEXAK8EPhgu+VIks5We3geQGY+Ajyh5VokSRU8M1GSCmdQS1LhDGpJKpxBLUmFM6glqXAGtSQVzqCWpMIZ1JJUOINakgpnUEtS4QxqSSqcQS1JhTOoJalwBrUkFc6glqTCGdSSVDiDWpIK1/RSXI+PiEMR8fmIOBIRv9p2YZKkkUaX4gLeAnwkM18RERcBj2uxJknSGWqDOiJ2A78O3AiQmd8Hvt9uWZKk0yIzt54Q8RzgIHAf8GxgDbh5fMHbM+etACsAnU5nb7/f3/B9hsMhCwsLtQWtHzuxjfInZ2nP7plst2lf2jCrXkN9v2fZlzZMotedXXD85ASKmZJp/U7Ny1pZXl5ey8xu1ViToO4Cnwaen5mfiYi3AN/JzD/f7Gu63W6urq5ueG4wGNDr9WqLXTxwuHZOG47esm8m223alzbMqtdQ3+9Z9qUNk+j1/qVT3LredG/l7E3rd2pe1kpEbBrUTT5MfBB4MDM/M358CPilSRUnSdpabVBn5v8CX4uIK8dPXcNoN4gkaQqa/h11E/Ce8REfDwC/315JkqQzNQrqzLwLqNx3Iklql2cmSlLhDGpJKpxBLUmFM6glqXAGtSQVzqCWpMIZ1JJUOINakgpnUEtS4QxqSSqcQS1JhTOoJalwBrUkFc6glqTCGdSSVDiDWpIKZ1BLUuEaXeElIo4C3wV+CJza7Eq5kqTJ286155cz8xutVSJJquSuD0kqXGRm/aSILwPfAhL4+8w8WDFnBVgB6HQ6e/v9/obx4XDIwsJC7bbWj51oVPi86OyCJ/307plse5a9Xtqz9c/cdL3sFJPodWcXHD85gWKmpO41npSqtTKrtf2T/MzLy8trm+1WbhrUezLzWEQ8CfgYcFNmfmKz+d1uN1dXVzc8NxgM6PV6tdtaPHC4ds482b90iptueNlMtj3LXh+9Zd+W403Xy04xiV7vXzrFrevb2Vs5W3Wv8aRUrZVZre2f5GeOiE2DutGuj8w8Nv73YeB24HmPuhpJ0rbUBnVEXBIRl56+D7wIuKftwiRJI03+juoAt0fE6fnvzcyPtFqVJOlHaoM6Mx8Anj2FWiRJFTw8T5IKZ1BLUuEMakkqnEEtSYUzqCWpcAa1JBXOoJakwhnUklQ4g1qSCmdQS1LhDGpJKpxBLUmFM6glqXAGtSQVzqCWpMIZ1JJUOINakgrXOKgj4oKIuDMiPtxmQZKkjbbzjvpm4EhbhUiSqjUK6oi4AtgHvK3dciRJZ4vMrJ8UcQj4a+BS4A2Z+ZKKOSvACkCn09nb7/c3jA+HQxYWFmq3tX7sRKPC50VnFxw/Oesqpm9pz+4tx5uul51iEut6p62Vutd4UqrWyqxy5Cf5mZeXl9cys1s1VnsV8oh4CfBwZq5FRG+zeZl5EDgI0O12s9fbOHUwGHD2c1VuPHC4ds482b90ilvXa1+GuXP0ht6W403Xy04xiXW909ZK3Ws8KVVrZVY50tbP3GTXx/OBl0bEUaAPXB0R726lGknSOWqDOjPfmJlXZOYicB3w8cx8ZeuVSZIAj6OWpOJta4dXZg6AQSuVSJIq+Y5akgpnUEtS4QxqSSqcQS1JhTOoJalwBrUkFc6glqTCGdSSVDiDWpIKZ1BLUuEMakkqnEEtSYUzqCWpcAa1JBXOoJakwhnUklQ4g1qSClcb1BFxcUT8V0TcHRH3RsRfTqMwSdJIk0txfQ+4OjOHEfFY4JMR8a+Z+emWa5Mk0SCoMzOB4fjhY8e3bLMoSdKPxSiHayZFXACsAT8LvDUz/7RizgqwAtDpdPb2+/0N48PhkIWFhdptrR870ajwedHZBcdPzrqK6Vvas3vL8abrZbt28vo6X9dKnZL6Ureut7K8vLyWmd2qsUZB/aPJEY8Hbgduysx7NpvX7XZzdXV1w3ODwYBer1e7jcUDhxvXMw/2L53i1vVtXQx+Lhy9Zd+W403Xy3bt5PV1vq6VOiX1pW5dbyUiNg3qbR31kZnfBu4Arn3U1UiStqXJUR9PHL+TJiJ2AS8EPt92YZKkkSZ/LzwZeOd4P/VjgPdl5ofbLUuSdFqToz4+Bzx3CrVIkip4ZqIkFc6glqTCGdSSVDiDWpIKZ1BLUuEMakkqnEEtSYUzqCWpcAa1JBXOoJakwhnUklQ4g1qSCmdQS1LhDGpJKpxBLUmFM6glqXAGtSQVrsk1E58SEXdExH0RcW9E3DyNwiRJI02umXgK2J+Zn42IS4G1iPhYZt7Xcm2SJBq8o87M/8nMz47vfxc4AuxpuzBJ0khkZvPJEYvAJ4BnZeZ3zhpbAVYAOp3O3n6/v+Frh8MhCwsLtdtYP3aicT3zoLMLjp+cdRXTt7Rn95bjTdfLdu3k9XW+rpU6JfWlbl1vZXl5eS0zu1VjjYM6IhaA/wDenJkf3Gput9vN1dXVDc8NBgN6vV7tdhYPHG5Uz7zYv3SKW9eb7IGaL0dv2bfleNP1sl07eX2dr2ulTkl9qVvXW4mITYO60VEfEfFY4APAe+pCWpI0WU2O+gjgH4Ejmfk37ZckSTpTk3fUzwdeBVwdEXeNby9uuS5J0ljtjp3M/CQQU6hFklTBMxMlqXAGtSQVzqCWpMIZ1JJUOINakgpnUEtS4QxqSSqcQS1JhTOoJalwBrUkFc6glqTCGdSSVDiDWpIKZ1BLUuEMakkqnEEtSYUzqCWpcE2umfj2iHg4Iu6ZRkGSpI2avKN+B3Bty3VIkjZRG9SZ+Qng/6ZQiySpQmRm/aSIReDDmfmsLeasACsAnU5nb7/f3zA+HA5ZWFio3db6sRO1c+ZJZxccPznrKspjX85lT6qV1JelPbsf9dcuLy+vZWa3aqz2KuRNZeZB4CBAt9vNXq+3YXwwGHD2c1VuPHB4UiXtCPuXTnHr+sRehrlhX85lT6qV1JejN/Ra+b4e9SFJhTOoJalwTQ7Puw34FHBlRDwYEa9uvyxJ0mm1O3Yy8/ppFCJJquauD0kqnEEtSYUzqCWpcAa1JBXOoJakwhnUklQ4g1qSCmdQS1LhDGpJKpxBLUmFM6glqXAGtSQVzqCWpMIZ1JJUOINakgpnUEtS4QxqSSpco6COiGsj4v6I+GJEHGi7KEnSjzW5ZuIFwFuB3wSuAq6PiKvaLkySNNLkHfXzgC9m5gOZ+X2gD7ys3bIkSadFZm49IeIVwLWZ+Zrx41cBv5yZrztr3gqwMn54JXD/Wd/qcuAbkyh6ztiXavblXPak2rz05amZ+cSqgdqrkDeVmQeBg5uNR8RqZnYntb15YV+q2Zdz2ZNq50Nfmuz6OAY85YzHV4yfkyRNQZOg/m/gGRHxtIi4CLgO+FC7ZUmSTqvd9ZGZpyLidcBHgQuAt2fmvY9iW5vuFjnP2Zdq9uVc9qTa3Pel9sNESdJseWaiJBXOoJakwk00qOtONY+IGyPi6xFx1/j2mkluv1RNTsGPiN+JiPsi4t6IeO+0a5yFBuvlb89YK1+IiG/Pos5pa9CXn4mIOyLizoj4XES8eBZ1TluDvjw1Iv593JNBRFwxizpbkZkTuTH6oPFLwNOBi4C7gavOmnMj8HeT2uZOuDXsyzOAO4HLxo+fNOu6S+jLWfNvYvRB9sxrn3VfGH149gfj+1cBR2dddyF9eT/we+P7VwPvmnXdk7pN8h21p5pXa9KX1wJvzcxvAWTmw1OucRa2u16uB26bSmWz1aQvCfzU+P5u4KEp1jcrTfpyFfDx8f07KsZ3rEkG9R7ga2c8fnD83Nl+e/ynyaGIeErF+Lxp0pdnAs+MiP+MiE9HxLVTq252mq4XIuKpwNP48S/hPGvSl78AXhkRDwL/wuivjXnXpC93Ay8f3/8t4NKIeMIUamvdtD9M/GdgMTN/EfgY8M4pb79UFzLa/dFj9M7xHyLi8TOtqCzXAYcy84ezLqQQ1wPvyMwrgBcD74oIDwyANwAviIg7gRcwOoN6LtbMJF/c2lPNM/Obmfm98cO3AXsnuP1SNTkF/0HgQ5n5g8z8MvAFRsE9z7bzXxNcx/mx2wOa9eXVwPsAMvNTwMWM/mOiedYkXx7KzJdn5nOBPxs/NxcfQE8yqGtPNY+IJ5/x8KXAkQluv1RNTsH/J0bvpomIyxntCnlgmkXOQKP/miAifg64DPjUlOublSZ9+SpwDUBE/DyjoP76VKucvib5cvkZf1m8EXj7lGtszcSCOjNPAadPNT8CvC8z742Iv4qIl46nvX58+NndwOsZHQUy1xr25aPANyPiPkYfgvxJZn5zNhVPR8O+wOgXsp/jj/LnXcO+7AdeO/49ug24cd7707AvPeD+iPgC0AHePJNiW+Ap5JJUOD+AkKTCGdSSVDiDWpIKZ1BLUuEMakkqnEEtSYUzqCWpcP8PLCeGkEJfpP4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Is 0.05 a good threshold? (check distribution- is there a clear cut-off?)\n",
    "max_cont_detect = not_detect_df[control_names].max(axis=1) #.hist()\n",
    "max_cont_detect_50 = max_cont_detect[max_cont_detect>0.5]#\n",
    "max_cont_detect_50.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1507d2110>"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAD4CAYAAAD1jb0+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAARkElEQVR4nO3df4zkdX3H8efbA4K9xUN7OLkAda8RTckZwZtQjEk7iz9yxQYwNUZSLVdp15pqTKXGq00jak01rZI2MWnPQu/aVFdKpVxAagjeSmwEuycnd0CtiGi5nlzR43S1pT1894/9XnNdd2a+Ozuz3/nsPR/JZOf7me/M933vzL32u5/5fucbmYkkqTzParoASdJgDHBJKpQBLkmFMsAlqVAGuCQV6rTV3NjGjRtzcnJyNTe5pB/+8IesX7++6TLGkr3pzf50Z296W0l/9u3b92RmnrN4fFUDfHJykrm5udXc5JJmZ2fpdDpNlzGW7E1v9qc7e9PbSvoTEd9aarz2FEpErIuI+yPi9mp5c0TcFxGPRMSnI+KMgSqTJA1kOXPg7wQePmn5I8ANmflC4Chw7TALkyT1VivAI+I84LXAX1bLAVwG3FKtshu4ahQFSpKWFnVOpY+IW4A/As4CfhfYDtxb7X0TEecDd2bmliWeOw1MA7Rara0zMzNDK35Q8/PzTExMNF3GWLI3vdmf7uxNbyvpz9TU1L7MbC8e7/shZkT8MnAkM/dFRGe5G87MncBOgHa7nePwIYcftnRnb3qzP93Zm95G0Z86R6G8ArgiIi4HzgSeA/wpcHZEnJaZx4HzgENDrUyS1FPfOfDM/L3MPC8zJ4E3Ap/PzF8F9gKvr1a7BrhtZFVKkn7CSs7EfA/wroh4BPhp4MbhlCRJqmNZJ/Jk5iwwW91/FLhk+CVJkupY1TMxJalJkzvuaGzbu7YN/2sG/DIrSSqUAS5JhTLAJalQBrgkFcoAl6RCGeCSVCgDXJIKZYBLUqEMcEkqlAEuSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1KhDHBJKlTfAI+IMyPiyxHx1Yh4MCLeX43viohvRsT+6nbR6MuVJJ1Q55JqTwOXZeZ8RJwOfDEi7qwee3dm3jK68iRJ3fQN8MxMYL5aPL265SiLkiT1Fwv53GeliHXAPuCFwMcz8z0RsQt4OQt76HcDOzLz6SWeOw1MA7Rara0zMzPDq35A8/PzTExMNF3GWLI3vdmf7krozYFDxxrb9uYN6wbuz9TU1L7MbC8erxXg/7dyxNnArcA7gO8C3wHOAHYC38jMD/R6frvdzrm5ueXUPRKzs7N0Op2myxhL9qY3+9NdCb1p+qr0g/YnIpYM8GUdhZKZTwF7gW2ZeTgXPA38FXDJQJVJkgZS5yiUc6o9byLi2cCrgX+JiE3VWABXAQdHWagk6f+rcxTKJmB3NQ/+LODmzLw9Ij4fEecAAewHfmuEdUqSFqlzFMoDwMVLjF82kookSbV4JqYkFcoAl6RCGeCSVCgDXJIKZYBLUqEMcEkqlAEuSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1KhDHBJKpQBLkmFMsAlqVAGuCQVqs41Mc+MiC9HxFcj4sGIeH81vjki7ouIRyLi0xFxxujLlSSdUGcP/Gngssx8KXARsC0iLgU+AtyQmS8EjgLXjq5MSdJifQM8F8xXi6dXtwQuA26pxnezcGV6SdIqqTUHHhHrImI/cAS4C/gG8FRmHq9WeRw4dzQlSpKWEplZf+WIs4FbgT8AdlXTJ0TE+cCdmblliedMA9MArVZr68zMzDDqXpH5+XkmJiaaLmMs2Zve7E93JfTmwKFjjW1784Z1A/dnampqX2a2F4+ftpwXycynImIv8HLg7Ig4rdoLPw841OU5O4GdAO12OzudznJrH7rZ2VnGoY5xZG96sz/dldCb7TvuaGzbu7atH3p/6hyFck61501EPBt4NfAwsBd4fbXaNcBtQ61MktRTnT3wTcDuiFjHQuDfnJm3R8RDwExE/CFwP3DjCOuUJC3SN8Az8wHg4iXGHwUuGUVRkqT+PBNTkgplgEtSoQxwSSqUAS5JhTLAJalQBrgkFcoAl6RCGeCSVCgDXJIKZYBLUqEMcEkqlAEuSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1Kh6lzU+PyI2BsRD0XEgxHxzmr8+og4FBH7q9vloy9XknRCnYsaHweuy8yvRMRZwL6IuKt67IbM/JPRlSdJ6qbORY0PA4er+z+IiIeBc0ddmCSpt8jM+itHTAL3AFuAdwHbge8DcyzspR9d4jnTwDRAq9XaOjMzs9KaV2x+fp6JiYmmyxhL9qY3+9NdCb05cOhYY9vevGHdwP2Zmpral5ntxeO1AzwiJoAvAB/KzM9ERAt4Ekjgg8CmzHxLr9dot9s5Nze37OKHbXZ2lk6n03QZY8ne9GZ/uiuhN5M77mhs27u2rR+4PxGxZIDXOgolIk4H/h7428z8DEBmPpGZz2Tmj4FPAJcMVJkkaSB1jkIJ4Ebg4cz82Enjm05a7XXAweGXJ0nqps5RKK8A3gwciIj91dh7gasj4iIWplAeA946kgolSUuqcxTKF4FY4qHPDr8cSVJdnokpSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1KhDHBJKpQBLkmFMsAlqVAGuCQVygCXpEIZ4JJUKANckgplgEtSoQxwSSqUAS5JhapzTczzI2JvRDwUEQ9GxDur8edFxF0R8fXq53NHX64k6YQ6e+DHgesy80LgUuC3I+JCYAdwd2ZeANxdLUuSVknfAM/Mw5n5ler+D4CHgXOBK4Hd1Wq7gatGVaQk6Sctaw48IiaBi4H7gFZmHq4e+g7QGmplkqSeIjPrrRgxAXwB+FBmfiYinsrMs096/Ghm/sQ8eERMA9MArVZr68zMzHAqX4H5+XkmJiaaLmMs2Zve7E93JfTmwKFjjW1784Z1A/dnampqX2a2F4/XCvCIOB24HfhcZn6sGvsa0MnMwxGxCZjNzBf3ep12u51zc3MD/QOGaXZ2lk6n03QZY8ne9GZ/uiuhN5M77mhs27u2rR+4PxGxZIDXOQolgBuBh0+Ed2UPcE11/xrgtoEqkyQN5LQa67wCeDNwICL2V2PvBT4M3BwR1wLfAt4wmhIlSUvpG+CZ+UUgujz8yuGWI0mqyzMxJalQBrgkFcoAl6RCGeCSVCgDXJIKZYBLUqEMcEkqlAEuSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1KhDHBJKpQBLkmFMsAlqVAGuCQVqs5FjW+KiCMRcfCksesj4lBE7K9ul4+2TEnSYnX2wHcB25YYvyEzL6punx1uWZKkfvoGeGbeA3xvFWqRJC1DZGb/lSImgdszc0u1fD2wHfg+MAdcl5lHuzx3GpgGaLVaW2dmZoZQ9srMz88zMTHRdBljyd70Zn+6K6E3Bw4da2zbmzesG7g/U1NT+zKzvXh80ABvAU8CCXwQ2JSZb+n3Ou12O+fm5pZX+QjMzs7S6XSaLmMs2Zve7E93JfRmcscdjW1717b1A/cnIpYM8IGOQsnMJzLzmcz8MfAJ4JKBqpIkDWygAI+ITSctvg442G1dSdJonNZvhYj4FNABNkbE48D7gE5EXMTCFMpjwFtHWKMkaQl9Azwzr15i+MYR1CJJWgbPxJSkQhngklQoA1ySCmWAS1KhDHBJKlTfo1AkadiaPCNyLXEPXJIKZYBLUqEMcEkqlAEuSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1KhDHBJKpQBLkmF6hvgEXFTRByJiIMnjT0vIu6KiK9XP5872jIlSYvV2QPfBWxbNLYDuDszLwDurpYlSauob4Bn5j3A9xYNXwnsru7vBq4acl2SpD4iM/uvFDEJ3J6ZW6rlpzLz7Op+AEdPLC/x3GlgGqDVam2dmZkZTuUrMD8/z8TERNNljCV705v96W45vTlw6NiIqxk/mzesG/i9MzU1tS8z24vHV3xBh8zMiOj6WyAzdwI7AdrtdnY6nZVucsVmZ2cZhzrGkb3pzf50t5zebD8FL+iwa9v6ob93Bj0K5YmI2ARQ/TwyvJIkSXUMGuB7gGuq+9cAtw2nHElSXXUOI/wU8CXgxRHxeERcC3wYeHVEfB14VbUsSVpFfefAM/PqLg+9csi1SFpFw76w8HUvOX5Kzm03yTMxJalQBrgkFcoAl6RCGeCSVCgDXJIKZYBLUqEMcEkqlAEuSYUywCWpUAa4JBXKAJekQhngklQoA1ySCmWAS1KhDHBJKpQBLkmFMsAlqVAruip9RDwG/AB4Bji+1GXvJUmjsaIAr0xl5pNDeB1J0jI4hSJJhYrMHPzJEd8EjgIJ/EVm7lxinWlgGqDVam2dmZkZeHvDMj8/z8TERNNljCV709ta6s+BQ8eG+nqtZ8MT/znUl1xTNm9YN/B7Z2pqat9SU9QrDfBzM/NQRDwfuAt4R2be0239drudc3NzA29vWGZnZ+l0Ok2XMZbsTW9rqT+juCr9Rw8MY1Z2bdq1bf3A752IWDLAVzSFkpmHqp9HgFuBS1byepKk+gYO8IhYHxFnnbgPvAY4OKzCJEm9reTvnRZwa0SceJ1PZuY/DqUqSVJfAwd4Zj4KvHSItUiSlsHDCCWpUAa4JBXKAJekQhngklQoA1ySClXMaVPDPGvsupccZ/syXu+xD792aNuWpGFxD1ySCmWAS1KhDHBJKpQBLkmFMsAlqVAGuCQVygCXpEIZ4JJUKANckgpVzJmYTRr2tQPrauoM0Kb+vSVY7lm8dXimrwblHrgkFcoAl6RCrSjAI2JbRHwtIh6JiB3DKkqS1N9Krkq/Dvg48EvAhcDVEXHhsAqTJPW2kj3wS4BHMvPRzPxvYAa4cjhlSZL6icwc7IkRrwe2ZeZvVMtvBn4+M9++aL1pYLpafDHwtcHLHZqNwJNNFzGm7E1v9qc7e9PbSvrzgsw8Z/HgyA8jzMydwM5Rb2c5ImIuM9tN1zGO7E1v9qc7e9PbKPqzkimUQ8D5Jy2fV41JklbBSgL8n4ELImJzRJwBvBHYM5yyJEn9DDyFkpnHI+LtwOeAdcBNmfng0CobrbGa0hkz9qY3+9Odvelt6P0Z+ENMSVKzPBNTkgplgEtSodZsgPc7zT8i3hURD0XEAxFxd0S8oIk6m1L3axAi4lciIiPilDk8rE5vIuIN1fvnwYj45GrX2KQa/7d+JiL2RsT91f+vy5uoswkRcVNEHImIg10ej4j4s6p3D0TEy1a0wcxcczcWPlT9BvCzwBnAV4ELF60zBfxUdf9twKebrnuc+lOtdxZwD3Av0G667nHpDXABcD/w3Gr5+U3XPWb92Qm8rbp/IfBY03WvYn9+AXgZcLDL45cDdwIBXArct5LtrdU98L6n+Wfm3sz8UbV4LwvHsZ8q6n4NwgeBjwD/tZrFNaxOb34T+HhmHgXIzCOrXGOT6vQngedU9zcA/76K9TUqM+8BvtdjlSuBv84F9wJnR8SmQbe3VgP8XODfTlp+vBrr5loWfiueKvr2p/rT7vzMPNWu7lDnvfMi4EUR8U8RcW9EbFu16ppXpz/XA2+KiMeBzwLvWJ3SirDcbOrplL8iT0S8CWgDv9h0LeMiIp4FfAzY3nAp4+o0FqZROiz85XZPRLwkM59qtKrxcTWwKzM/GhEvB/4mIrZk5o+bLmytWat74LVO84+IVwG/D1yRmU+vUm3joF9/zgK2ALMR8RgLc3V7TpEPMuu8dx4H9mTm/2TmN4F/ZSHQTwV1+nMtcDNAZn4JOJOFL3LSkL+CZK0GeN/T/CPiYuAvWAjvU2kOE/r0JzOPZebGzJzMzEkWPiO4IjPnmil3VdX5ioh/YGHvm4jYyMKUyqOrWWSD6vTn28ArASLi51gI8P9Y1SrH1x7g16qjUS4FjmXm4UFfbE1OoWSX0/wj4gPAXGbuAf4YmAD+LiIAvp2ZVzRW9Cqq2Z9TUs3efA54TUQ8BDwDvDszv9tc1aunZn+uAz4REb/Dwgea27M6BGOti4hPsfDLfWP1GcD7gNMBMvPPWfhM4HLgEeBHwK+vaHunSF8lac1Zq1MokrTmGeCSVCgDXJIKZYBLUqEMcEkqlAEuSYUywCWpUP8LzDM8MKfmdWsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "s_detect = not_detect_df['s']\n",
    "s_detect.hist()#[s_detect>0.5].hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check that contigs of known contam taxonomy are removed\n",
    "Lysobacter and Asaccharospora"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_contigs = tax.index\n",
    "lysobacter_contigs = tax[tax.genus=='Lysobacter'].index\n",
    "asaccharospora_contigs = tax[tax.genus=='Asaccharospora'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s</th>\n",
       "      <th>neg_ex_1</th>\n",
       "      <th>neg_field_1</th>\n",
       "      <th>neg_field_2</th>\n",
       "      <th>neg_field_bf</th>\n",
       "      <th>pos_ex_1</th>\n",
       "      <th>pos_ex_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_27354_split_00001</td>\n",
       "      <td>1.452609</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_420504_split_00001</td>\n",
       "      <td>2.112500</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_352364_split_00001</td>\n",
       "      <td>2.722191</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_506679_split_00001</td>\n",
       "      <td>2.806985</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_106197_split_00001</td>\n",
       "      <td>2.893726</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_311382_split_00001</td>\n",
       "      <td>2.978178</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_445879_split_00001</td>\n",
       "      <td>3.023963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_110577_split_00001</td>\n",
       "      <td>3.162200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_318329_split_00001</td>\n",
       "      <td>3.281389</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_315591_split_00001</td>\n",
       "      <td>3.348214</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_300862_split_00001</td>\n",
       "      <td>3.374793</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_248235_split_00001</td>\n",
       "      <td>3.809392</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_186487_split_00001</td>\n",
       "      <td>3.941547</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_231693_split_00001</td>\n",
       "      <td>4.065385</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_47226_split_00001</td>\n",
       "      <td>4.175610</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_74186_split_00001</td>\n",
       "      <td>4.416490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_10616_split_00001</td>\n",
       "      <td>4.920611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_434404_split_00001</td>\n",
       "      <td>5.378057</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_197792_split_00001</td>\n",
       "      <td>7.425892</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_394938_split_00001</td>\n",
       "      <td>8.075840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_425963_split_00001</td>\n",
       "      <td>12.314000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_122987_split_00001</td>\n",
       "      <td>17.401329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    s  neg_ex_1  neg_field_1  neg_field_2  \\\n",
       "split                                                                       \n",
       "KNLK_58_27354_split_00001    1.452609       0.0          0.0          0.0   \n",
       "KNLK_58_420504_split_00001   2.112500       0.0          0.0          0.0   \n",
       "KNLK_58_352364_split_00001   2.722191       0.0          0.0          0.0   \n",
       "KNLK_58_506679_split_00001   2.806985       0.0          0.0          0.0   \n",
       "KNLK_58_106197_split_00001   2.893726       0.0          0.0          0.0   \n",
       "KNLK_58_311382_split_00001   2.978178       0.0          0.0          0.0   \n",
       "KNLK_58_445879_split_00001   3.023963       0.0          0.0          0.0   \n",
       "KNLK_58_110577_split_00001   3.162200       0.0          0.0          0.0   \n",
       "KNLK_58_318329_split_00001   3.281389       0.0          0.0          0.0   \n",
       "KNLK_58_315591_split_00001   3.348214       0.0          0.0          0.0   \n",
       "KNLK_58_300862_split_00001   3.374793       0.0          0.0          0.0   \n",
       "KNLK_58_248235_split_00001   3.809392       0.0          0.0          0.0   \n",
       "KNLK_58_186487_split_00001   3.941547       0.0          0.0          0.0   \n",
       "KNLK_58_231693_split_00001   4.065385       0.0          0.0          0.0   \n",
       "KNLK_58_47226_split_00001    4.175610       0.0          0.0          0.0   \n",
       "KNLK_58_74186_split_00001    4.416490       0.0          0.0          0.0   \n",
       "KNLK_58_10616_split_00001    4.920611       0.0          0.0          0.0   \n",
       "KNLK_58_434404_split_00001   5.378057       0.0          0.0          0.0   \n",
       "KNLK_58_197792_split_00001   7.425892       0.0          0.0          0.0   \n",
       "KNLK_58_394938_split_00001   8.075840       0.0          0.0          0.0   \n",
       "KNLK_58_425963_split_00001  12.314000       0.0          0.0          0.0   \n",
       "KNLK_58_122987_split_00001  17.401329       0.0          0.0          0.0   \n",
       "\n",
       "                            neg_field_bf  pos_ex_1  pos_ex_2  \n",
       "split                                                         \n",
       "KNLK_58_27354_split_00001            0.0       0.0       0.0  \n",
       "KNLK_58_420504_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_352364_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_506679_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_106197_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_311382_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_445879_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_110577_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_318329_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_315591_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_300862_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_248235_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_186487_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_231693_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_47226_split_00001            0.0       0.0       0.0  \n",
       "KNLK_58_74186_split_00001            0.0       0.0       0.0  \n",
       "KNLK_58_10616_split_00001            0.0       0.0       0.0  \n",
       "KNLK_58_434404_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_197792_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_394938_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_425963_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_122987_split_00001           0.0       0.0       0.0  "
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# which Lysobacter contigs were missed by the decontamination? 22. All of them had 0x coverage in controls.\n",
    "missed_lysobacter = list(set(lysobacter_contigs) - set(possible_contams_contigs))\n",
    "cov_means[cov_means.index.isin(missed_lysobacter)].sort_values('s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# which Asaccharospora contigs were missed by the decontamination? None!\n",
    "len(set(asaccharospora_contigs) - set(possible_contams_contigs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s</th>\n",
       "      <th>neg_ex_1</th>\n",
       "      <th>neg_field_1</th>\n",
       "      <th>neg_field_2</th>\n",
       "      <th>neg_field_bf</th>\n",
       "      <th>pos_ex_1</th>\n",
       "      <th>pos_ex_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_138203_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_171115_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_210574_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_47619_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              s  neg_ex_1  neg_field_1  neg_field_2  \\\n",
       "split                                                                 \n",
       "KNLK_58_138203_split_00001  0.0       0.0          0.0          0.0   \n",
       "KNLK_58_171115_split_00001  0.0       0.0          0.0          0.0   \n",
       "KNLK_58_210574_split_00001  0.0       0.0          0.0          0.0   \n",
       "KNLK_58_47619_split_00001   0.0       0.0          0.0          0.0   \n",
       "\n",
       "                            neg_field_bf  pos_ex_1  pos_ex_2  \n",
       "split                                                         \n",
       "KNLK_58_138203_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_171115_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_210574_split_00001           0.0       0.0       0.0  \n",
       "KNLK_58_47619_split_00001            0.0       0.0       0.0  "
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are 4 contigs that have 0 coverage in all controls and the sample (after filtering) - toss these?\n",
    "cov_means[(cov_means == 0).all(axis = 'columns')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_ex_1</th>\n",
       "      <th>neg_field_1</th>\n",
       "      <th>neg_field_2</th>\n",
       "      <th>neg_field_bf</th>\n",
       "      <th>pos_ex_1</th>\n",
       "      <th>pos_ex_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>split</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>KNLK_58_100007_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_100008_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_100009_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_100013_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_100031_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_99985_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_99986_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_99988_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_99991_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>KNLK_58_99997_split_00001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106339 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            neg_ex_1  neg_field_1  neg_field_2  neg_field_bf  \\\n",
       "split                                                                          \n",
       "KNLK_58_100007_split_00001       0.0          0.0          0.0           0.0   \n",
       "KNLK_58_100008_split_00001       0.0          0.0          0.0           0.0   \n",
       "KNLK_58_100009_split_00001       0.0          0.0          0.0           0.0   \n",
       "KNLK_58_100013_split_00001       0.0          0.0          0.0           0.0   \n",
       "KNLK_58_100031_split_00001       0.0          0.0          0.0           0.0   \n",
       "...                              ...          ...          ...           ...   \n",
       "KNLK_58_99985_split_00001        0.0          0.0          0.0           0.0   \n",
       "KNLK_58_99986_split_00001        0.0          0.0          0.0           0.0   \n",
       "KNLK_58_99988_split_00001        0.0          0.0          0.0           0.0   \n",
       "KNLK_58_99991_split_00001        0.0          0.0          0.0           0.0   \n",
       "KNLK_58_99997_split_00001        0.0          0.0          0.0           0.0   \n",
       "\n",
       "                            pos_ex_1  pos_ex_2  \n",
       "split                                           \n",
       "KNLK_58_100007_split_00001       0.0       0.0  \n",
       "KNLK_58_100008_split_00001       0.0       0.0  \n",
       "KNLK_58_100009_split_00001       0.0       0.0  \n",
       "KNLK_58_100013_split_00001       0.0       0.0  \n",
       "KNLK_58_100031_split_00001       0.0       0.0  \n",
       "...                              ...       ...  \n",
       "KNLK_58_99985_split_00001        0.0       0.0  \n",
       "KNLK_58_99986_split_00001        0.0       0.0  \n",
       "KNLK_58_99988_split_00001        0.0       0.0  \n",
       "KNLK_58_99991_split_00001        0.0       0.0  \n",
       "KNLK_58_99997_split_00001        0.0       0.0  \n",
       "\n",
       "[106339 rows x 6 columns]"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are 106339 contigs that have 0 coverage in all controls (after filtering)\n",
    "cov_means[(cov_means[control_names] == 0).all(axis = 'columns')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>split</th>\n",
       "      <th>bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>KNLK_58_100189_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>KNLK_58_100346_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KNLK_58_10037_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>KNLK_58_100456_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KNLK_58_100490_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1673</td>\n",
       "      <td>KNLK_58_99127_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1674</td>\n",
       "      <td>KNLK_58_99127_split_00002</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1675</td>\n",
       "      <td>KNLK_58_99370_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1676</td>\n",
       "      <td>KNLK_58_99632_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1677</td>\n",
       "      <td>KNLK_58_99656_split_00001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1678 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           split     bin\n",
       "0     KNLK_58_100189_split_00001  contam\n",
       "1     KNLK_58_100346_split_00001  contam\n",
       "2      KNLK_58_10037_split_00001  contam\n",
       "3     KNLK_58_100456_split_00001  contam\n",
       "4     KNLK_58_100490_split_00001  contam\n",
       "...                          ...     ...\n",
       "1673   KNLK_58_99127_split_00001  contam\n",
       "1674   KNLK_58_99127_split_00002  contam\n",
       "1675   KNLK_58_99370_split_00001  contam\n",
       "1676   KNLK_58_99632_split_00001  contam\n",
       "1677   KNLK_58_99656_split_00001  contam\n",
       "\n",
       "[1678 rows x 2 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "collection = pd.DataFrame(possible_contams.index)\n",
    "collection['bin'] = 'contam'\n",
    "collection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contigs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>KNLK_58_100346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>KNLK_58_100456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KNLK_58_100490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>KNLK_58_100836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KNLK_58_100999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1613</td>\n",
       "      <td>KNLK_58_87401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1614</td>\n",
       "      <td>KNLK_58_90394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1615</td>\n",
       "      <td>KNLK_58_92561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1616</td>\n",
       "      <td>KNLK_58_93295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1617</td>\n",
       "      <td>KNLK_58_9817</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1618 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             contigs\n",
       "0     KNLK_58_100346\n",
       "1     KNLK_58_100456\n",
       "2     KNLK_58_100490\n",
       "3     KNLK_58_100836\n",
       "4     KNLK_58_100999\n",
       "...              ...\n",
       "1613   KNLK_58_87401\n",
       "1614   KNLK_58_90394\n",
       "1615   KNLK_58_92561\n",
       "1616   KNLK_58_93295\n",
       "1617    KNLK_58_9817\n",
       "\n",
       "[1618 rows x 1 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# trying to get contig column for testing\n",
    "t = possible_contams.reset_index()\n",
    "t['contig'] = t['split'].str.extract(r'(.+)_split_\\d+')\n",
    "t= t.drop(columns='split')\n",
    "t = t.drop_duplicates()\n",
    "t = t.set_index('contig')\n",
    "\n",
    "iscontamm = t.reset_index().melt(id_vars=['contig'],\n",
    "                          var_name='control',\n",
    "                          value_name='contamination')\n",
    "p = pd.DataFrame(iscontamm[iscontamm.contamination!='noncontam'].contig.unique(), columns=['contigs'])\n",
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "# questions\n",
    "# what fraction of negative controls is from positive control contam\n",
    "# what should rule be: 1) if any control says the contig is a contam, then it is contam (what about cross-contam of samples into just one control?)\n",
    "#                      2) if any control says contig is noncontam, then it is noncontam (some controls won't have enough info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "## import contig table\n",
    "# contig_data = pd.read_csv(contig_data_file, sep='\\t')\n",
    "# contig_data = contig_data.set_index('contig')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "possible_contams = pd.DataFrame(iscontamm[iscontamm.contamination!=2].contig.unique(), columns=['contigs'])\n",
    "possible_contams['bin']='contam'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contigs</th>\n",
       "      <th>bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>KNLK_23_314861</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>KNLK_23_63001</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KNLK_23_223930</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>KNLK_23_279903</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KNLK_23_237933</td>\n",
       "      <td>contam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          contigs     bin\n",
       "0  KNLK_23_314861  contam\n",
       "1   KNLK_23_63001  contam\n",
       "2  KNLK_23_223930  contam\n",
       "3  KNLK_23_279903  contam\n",
       "4  KNLK_23_237933  contam"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possible_contams.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "possible_contams.to_csv(out_file, header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "profileDB = '~/data/awtp2_metagenomics/test_decontamination/KNLK_58_merged_profiles_filtered/PROFILE.db'\n",
    "contigsDB = '~/data/awtp2_metagenomics/test_decontamination/KNLK_58_contigs.db'\n",
    "name = 'AWTP_2_RO2_bulk_3'\n",
    "mcov_cmd = f'anvi-export-table --table mean_coverage_Q2Q3_splits {profileDB} -o {name}_coverage_Q2Q3_splits.txt'\n",
    "scov_cmd = f'anvi-export-table --table std_coverage_splits {profileDB} -o {name}_std_coverage_splits.txt'\n",
    "dcov_cmd = f'anvi-export-table -t detection_splits {profileDB} -o {name}_detection_splits.txt'\n",
    "tax_cmd = f'anvi-export-splits-taxonomy -c {contigsDB} -o {name}_taxonomy_splits.txt'\n",
    "all_cmd = [mcov_cmd, scov_cmd, dcov_cmd, tax_cmd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'anvi-export-table': 'anvi-export-table'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-163-defeeb50d270>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mshlex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmcov_cmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;31m#subprocess.run(['anvi-export-table', '--table mean_coverage_Q2Q3_splits {profileDB} -o {name}_coverage_Q2Q3_splits.txt])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.7.5/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    486\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'stderr'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPIPE\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    487\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 488\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    489\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    490\u001b[0m             \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.7.5/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, encoding, errors, text)\u001b[0m\n\u001b[1;32m    798\u001b[0m                                 \u001b[0mc2pread\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc2pwrite\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    799\u001b[0m                                 \u001b[0merrread\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrwrite\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 800\u001b[0;31m                                 restore_signals, start_new_session)\n\u001b[0m\u001b[1;32m    801\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    802\u001b[0m             \u001b[0;31m# Cleanup if the child failed starting.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.pyenv/versions/3.7.5/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36m_execute_child\u001b[0;34m(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, start_new_session)\u001b[0m\n\u001b[1;32m   1549\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0merrno_num\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0merrno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mENOENT\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1550\u001b[0m                             \u001b[0merr_msg\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m': '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mrepr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1551\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mchild_exception_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno_num\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_msg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1552\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mchild_exception_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1553\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'anvi-export-table': 'anvi-export-table'"
     ]
    }
   ],
   "source": [
    "#for i in all_cmd:\n",
    "#    subprocess.run(shlex.split(i))\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
