{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import libraries and metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "wd = '/Users/rosekantor/data/awtp2_metagenomics'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "tab = pd.read_csv(f'{wd}/metagenomics_sample_table.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>read_names</th>\n",
       "      <th>sample_16S</th>\n",
       "      <th>order</th>\n",
       "      <th>band_in_gel_amplicons</th>\n",
       "      <th>sample_code_partial</th>\n",
       "      <th>Sequal_prep_elute_pooled</th>\n",
       "      <th>Duplicate.</th>\n",
       "      <th>need_optimization.</th>\n",
       "      <th>Has_Other_pos_sample</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_or_control</th>\n",
       "      <th>location_code</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sampler_name</th>\n",
       "      <th>Ext_batch</th>\n",
       "      <th>Extraction_Date</th>\n",
       "      <th>Ext_num_unique</th>\n",
       "      <th>Microconcentrated.</th>\n",
       "      <th>DNA_Ext_conc_ngperuL</th>\n",
       "      <th>kit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>KNLK_1</td>\n",
       "      <td>KNLK_1_S87</td>\n",
       "      <td>KNLK_SD176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_biofilm_control_1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>control</td>\n",
       "      <td>RO_BF_control</td>\n",
       "      <td>biofilm</td>\n",
       "      <td>Rose_Kantor</td>\n",
       "      <td>RO1</td>\n",
       "      <td>9/13/2018</td>\n",
       "      <td>A58</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.059</td>\n",
       "      <td>Powersoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>KNLK_10</td>\n",
       "      <td>KNLK_10_S73</td>\n",
       "      <td>KNLK_SD021</td>\n",
       "      <td>73.0</td>\n",
       "      <td>+</td>\n",
       "      <td>DWDS_F_282B_bulk_1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>DWDS_282B</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Lauren_Kennedy</td>\n",
       "      <td>7</td>\n",
       "      <td>7/5/2018</td>\n",
       "      <td>42</td>\n",
       "      <td>Y</td>\n",
       "      <td>24.600</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KNLK_11</td>\n",
       "      <td>KNLK_11_S81</td>\n",
       "      <td>KNLK_SD028</td>\n",
       "      <td>81.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_BAC_bulk_5</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>BAC</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>2</td>\n",
       "      <td>6/16/2018</td>\n",
       "      <td>15</td>\n",
       "      <td>N</td>\n",
       "      <td>169.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>KNLK_12</td>\n",
       "      <td>KNLK_12_S2</td>\n",
       "      <td>KNLK_SD178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>SDS_F_3_bulk_3</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>Loop_3</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>unk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>W7</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.154</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KNLK_13</td>\n",
       "      <td>KNLK_13_S10</td>\n",
       "      <td>KNLK_SD177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_biofilm_sep_2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_BF_separator</td>\n",
       "      <td>biofilm</td>\n",
       "      <td>Rose_Kantor</td>\n",
       "      <td>19</td>\n",
       "      <td>9/27/2018</td>\n",
       "      <td>A69</td>\n",
       "      <td>N</td>\n",
       "      <td>0.071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id   read_names  sample_16S  order band_in_gel_amplicons  \\\n",
       "0    KNLK_1   KNLK_1_S87  KNLK_SD176    NaN                     +   \n",
       "1   KNLK_10  KNLK_10_S73  KNLK_SD021   73.0                     +   \n",
       "2   KNLK_11  KNLK_11_S81  KNLK_SD028   81.0                     +   \n",
       "3   KNLK_12   KNLK_12_S2  KNLK_SD178    NaN                     +   \n",
       "4   KNLK_13  KNLK_13_S10  KNLK_SD177    NaN                     +   \n",
       "\n",
       "            sample_code_partial Sequal_prep_elute_pooled  Duplicate.  \\\n",
       "0  AWTP_2_RO2_biofilm_control_1                        N       False   \n",
       "1            DWDS_F_282B_bulk_1                        N       False   \n",
       "2             AWTP_2_BAC_bulk_5                        N       False   \n",
       "3                SDS_F_3_bulk_3                        N       False   \n",
       "4      AWTP_2_RO2_biofilm_sep_2                        N       False   \n",
       "\n",
       "  need_optimization. Has_Other_pos_sample  ... sample_or_control  \\\n",
       "0                  N                    N  ...           control   \n",
       "1                  N                  NaN  ...            sample   \n",
       "2                  N                  NaN  ...            sample   \n",
       "3                  N                  NaN  ...            sample   \n",
       "4                  N                  NaN  ...            sample   \n",
       "\n",
       "     location_code sample_type    sampler_name  Ext_batch  Extraction_Date  \\\n",
       "0    RO_BF_control     biofilm     Rose_Kantor        RO1        9/13/2018   \n",
       "1        DWDS_282B        DEUF  Lauren_Kennedy          7         7/5/2018   \n",
       "2              BAC        DEUF    Scott_Miller          2        6/16/2018   \n",
       "3           Loop_3        DEUF    Scott_Miller        unk              NaN   \n",
       "4  RO_BF_separator     biofilm     Rose_Kantor         19        9/27/2018   \n",
       "\n",
       "   Ext_num_unique  Microconcentrated. DNA_Ext_conc_ngperuL            kit  \n",
       "0             A58                   Y                0.059  Powersoil Pro  \n",
       "1              42                   Y               24.600  PowerSoil Pro  \n",
       "2              15                   N              169.000  PowerSoil Pro  \n",
       "3              W7         supernatant                0.154            NaN  \n",
       "4             A69                   N                0.071            NaN  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change project_type for 'Loop_MWTP_inf' to be Full_Scale so that we can group by project_type later and MWTP will be with DWDS\n",
    "tab.loc[tab['location_code']=='Loop_MWTP_inf', ['project_type']] = 'Full_Scale'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DPWF', 'ECAWPC', 'Experimental', 'Full_Scale', 'Pipe_Loop'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(tab.project_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read trimming and QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preliminary FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# did this on the awtp samples, not everything.  Will check everything after trimming rather than before.\n",
    "awtp =  tab[tab.project_type.isin(['DPWF', 'ECAWPC'])]\n",
    "\n",
    "with open(f'{wd}/read_fastqc.sh', 'w') as f:\n",
    "    for i in awtp.read_names:\n",
    "        f.write(f'/opt/bin/bio/FastQC/fastqc -t 2 {i}.1.fastq.gz {i}.2.fastq.gz\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manual examination of fastQC showed forward reads have adapter contamination at 3' end. Reverse reads have small amounts of low-quality sequence as indicated by over-represented sequences that were strings of G's. Trimming dealt with both of these successfully.\n",
    "\n",
    "A few samples look like they failed sequencing (high adapter contamination, weird per-base content and weird quality profiles)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trim bbmap and sickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "bb = '/opt/bin/bio/bbmap/bbduk.sh'\n",
    "sickle = '/opt/bin/bio/sickle'\n",
    "fastqc = '/opt/bin/bio/FastQC/fastqc'\n",
    "adapters = '/opt/bin/bio/bbmap/resources/adapters.fa'\n",
    "phiX = '/opt/bin/bio/bbmap/resources/phix174_ill.ref.fa.gz'\n",
    "settings = '-Xmx5g k=23 mink=11 hdist=1 tbo tpe t=2'\n",
    "\n",
    "# group by project type, 'group' will be the name of the project type, 'df' will be a filtered pandas dataframe for just that project\n",
    "# iterate through each filtered dataframe and make all the reads processing commands\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    with open(f'{wd}/workflows/read_trim_{group}.sh', 'w') as f:\n",
    "        for row in df.itertuples():\n",
    "            r = row.read_names\n",
    "            s = row.sample_id\n",
    "            trimR = f'{bb} ref={adapters} {settings} ktrim=r ftm=5 -in1={r}.1.fastq.gz -in2={r}.2.fastq.gz -out1={s}.1.noadapt.fastq.gz -out2={s}.2.noadapt.fastq.gz'\n",
    "            #filtphiX = f'{bb} -Xmx5g t=2 k=31 hdist=1 -in1={s}.1.noadapt.fastq.gz -in2={s}.2.noadapt.fastq.gz out1={s}.1.nophix.fastq.gz out2={s}.2.nophix.fastq.gz ref={phiX} stats={s}_stats_phix.txt'\n",
    "            # based on testing, reads had no phiX, so this command isn't necessary\n",
    "            qtrim = f'sickle pe -l 75 -f {s}.1.noadapt.fastq.gz -r {s}.2.noadapt.fastq.gz -t sanger -o {s}.PE.1.fastq -p {s}.PE.2.fastq -s {s}.SR.fastq'\n",
    "            clean = f'rm {s}.1.noadapt.fastq.gz {s}.2.noadapt.fastq.gz {s}.SR.fastq'\n",
    "            qc = f'{fastqc} -t 2 {s}.PE.1.fastq {s}.PE.2.fastq'\n",
    "            gz1 = f'pigz -p 2 {s}.PE.1.fastq'\n",
    "            gz2 = f'pigz -p 2 {s}.PE.2.fastq'\n",
    "\n",
    "            cmd = [trimR, qtrim, clean, qc, gz1, gz2]\n",
    "            f.write('\\n'.join(cmd) + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>read_names</th>\n",
       "      <th>sample_16S</th>\n",
       "      <th>order</th>\n",
       "      <th>band_in_gel_amplicons</th>\n",
       "      <th>sample_code_partial</th>\n",
       "      <th>Sequal_prep_elute_pooled</th>\n",
       "      <th>Duplicate.</th>\n",
       "      <th>need_optimization.</th>\n",
       "      <th>Has_Other_pos_sample</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_or_control</th>\n",
       "      <th>location_code</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sampler_name</th>\n",
       "      <th>Ext_batch</th>\n",
       "      <th>Extraction_Date</th>\n",
       "      <th>Ext_num_unique</th>\n",
       "      <th>Microconcentrated.</th>\n",
       "      <th>DNA_Ext_conc_ngperuL</th>\n",
       "      <th>kit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>28</td>\n",
       "      <td>KNLK_35</td>\n",
       "      <td>KNLK_35_S12</td>\n",
       "      <td>KNLK_SD001</td>\n",
       "      <td>12.0</td>\n",
       "      <td>+</td>\n",
       "      <td>EXP_0_field_bulk</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>control</td>\n",
       "      <td>field_blank</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Lauren_Kennedy</td>\n",
       "      <td>19</td>\n",
       "      <td>9/8/2018</td>\n",
       "      <td>183</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.126</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33</td>\n",
       "      <td>KNLK_4</td>\n",
       "      <td>KNLK_4_S25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "      <td>-</td>\n",
       "      <td>EXP_0_field_bulk</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id   read_names  sample_16S  order band_in_gel_amplicons  \\\n",
       "28   KNLK_35  KNLK_35_S12  KNLK_SD001   12.0                     +   \n",
       "33    KNLK_4   KNLK_4_S25         NaN   25.0                     -   \n",
       "\n",
       "   sample_code_partial Sequal_prep_elute_pooled  Duplicate.  \\\n",
       "28    EXP_0_field_bulk                        N       False   \n",
       "33    EXP_0_field_bulk                      NaN       False   \n",
       "\n",
       "   need_optimization. Has_Other_pos_sample  ... sample_or_control  \\\n",
       "28                  N                    N  ...           control   \n",
       "33                  N                    N  ...               NaN   \n",
       "\n",
       "   location_code sample_type    sampler_name  Ext_batch  Extraction_Date  \\\n",
       "28   field_blank        DEUF  Lauren_Kennedy         19         9/8/2018   \n",
       "33           NaN         NaN             NaN        NaN              NaN   \n",
       "\n",
       "    Ext_num_unique  Microconcentrated. DNA_Ext_conc_ngperuL            kit  \n",
       "28             183                   Y                0.126  PowerSoil Pro  \n",
       "33             NaN                 NaN                  NaN            NaN  \n",
       "\n",
       "[2 rows x 46 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#len(tab.sample_code_partial)\n",
    "tab[tab.sample_code_partial.duplicated()==True]\n",
    "tab[tab.sample_code_partial=='EXP_0_field_bulk']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MASH within-project all-vs-all beta diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#using MASH v2.0 (no -I option)\n",
    "mash = '/opt/bin/bio/mash'\n",
    "\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    with open(f'{wd}/workflows/read_mash_{group}.sh', 'w') as f:\n",
    "        for row in df.itertuples():\n",
    "            s = row.sample_id\n",
    "            concat = f'cat {s}.PE.1.fastq.gz {s}.PE.2.fastq.gz > {s}'\n",
    "            sketch = f'{mash} sketch -m 2 -r {s} -s 10000'\n",
    "            clean = f'rm {s}'\n",
    "            cmd = [concat, sketch, clean]\n",
    "            f.write('\\n'.join(cmd) + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using MASH v2.2 (with -I option that allows you to provide a sample name with stdin)\n",
    "# using stdin allows you to concatenate fwd and rev reads and pipe directly into mash without making intermediate files.\n",
    "# this is denoted with a dash (\"-\") after the -r parameter instead of a reads name\n",
    "\n",
    "mash = '/opt/bin/bio/mash'\n",
    "\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    with open(f'{wd}/workflows/read_mash_{group}.sh', 'w') as f:\n",
    "        for row in df.itertuples():\n",
    "            s = row.sample_id\n",
    "            cmd = f'cat {s}.PE.1.fastq.gz {s}.PE.2.fastq.gz | {mash} sketch -m 2 -r - -I {s} -s 10000'\n",
    "            f.write(cmd + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mash dist commands:\n",
    "`mash paste` combines multiple sketches into a single sketch.  The first arg is output name, followed by a list of all the sketch files you want to combine\n",
    "\n",
    "Command: `mash paste awtp2.msh *msh`\n",
    "\n",
    "`mash dist` can sketch on the fly or take a sketch as input.  Because we are doing all-vs-all we use the same msh file as the query and reference.\n",
    "\n",
    "Command: `mash dist awtp2.msh awtp2.msh`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "megahit = '/opt/bin/bio/megahit'\n",
    "assem_dir = '/data2/other/knelson/assembly'\n",
    "reads_dir = '/data2/other/knelson/raw.d/trimmed_reads'\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    with open(f'{wd}/workflows/megahit_{group}.sh', 'w') as f:\n",
    "        for row in df.itertuples():\n",
    "            s = row.sample_id\n",
    "            assem_name = row.sample_code_partial\n",
    "            assem_cmd = f'{megahit} -t 48 ' \\\n",
    "                        f'-1 {reads_dir}/{s}.PE.1.fastq.gz ' \\\n",
    "                        f'-2 {reads_dir}/{s}.PE.1.fastq.gz ' \\\n",
    "                        f'-o {assem_dir}/{assem_name}'\n",
    "            qcmd = f'echo \"{assem_cmd}\" | qsub -V -pe smp 48 -N {s}megahit'\n",
    "            f.write(qcmd + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42.050146383333335\n",
      "256.03558838333333\n"
     ]
    }
   ],
   "source": [
    "# ASSEMBLY TIME:\n",
    "\n",
    "# KNLK_16megahit assembly time (min):\n",
    "print(2523.008783/60)\n",
    "# KNLK_76 assembly time\n",
    "print(15362.135303/60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59.779999999999994\n",
      "63.17000000000001\n"
     ]
    }
   ],
   "source": [
    "# ASSEMBLY QUALITY:\n",
    "\n",
    "# megahit bps on contigs > 1kb\n",
    "print(100 - 22.37 - 17.85)\n",
    "\n",
    "# idba_ud bp on contigs > 1kb\n",
    "print(100 -  0.02 - 15.88 - 20.93)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>read_names</th>\n",
       "      <th>sample_16S</th>\n",
       "      <th>order</th>\n",
       "      <th>band_in_gel_amplicons</th>\n",
       "      <th>sample_code_partial</th>\n",
       "      <th>Sequal_prep_elute_pooled</th>\n",
       "      <th>Duplicate.</th>\n",
       "      <th>need_optimization.</th>\n",
       "      <th>Has_Other_pos_sample</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_or_control</th>\n",
       "      <th>location_code</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sampler_name</th>\n",
       "      <th>Ext_batch</th>\n",
       "      <th>Extraction_Date</th>\n",
       "      <th>Ext_num_unique</th>\n",
       "      <th>Microconcentrated.</th>\n",
       "      <th>DNA_Ext_conc_ngperuL</th>\n",
       "      <th>kit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>KNLK_9</td>\n",
       "      <td>KNLK_9_S65</td>\n",
       "      <td>KNLK_SD147</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_BAC_bulk_2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>BAC</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>2</td>\n",
       "      <td>6/16/2018</td>\n",
       "      <td>13</td>\n",
       "      <td>N</td>\n",
       "      <td>17.500</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>KNLK_32</td>\n",
       "      <td>KNLK_32_S75</td>\n",
       "      <td>KNLK_SD023</td>\n",
       "      <td>75.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_BAC_bulk_3</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>BAC</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>Test_March11</td>\n",
       "      <td>3/11/2018</td>\n",
       "      <td>W5</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>27.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>73</td>\n",
       "      <td>KNLK_76</td>\n",
       "      <td>KNLK_76_S79</td>\n",
       "      <td>KNLK_SD160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_BAC_bulk_4</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>BAC</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>2</td>\n",
       "      <td>6/16/2018</td>\n",
       "      <td>14</td>\n",
       "      <td>N</td>\n",
       "      <td>85.600</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>KNLK_11</td>\n",
       "      <td>KNLK_11_S81</td>\n",
       "      <td>KNLK_SD028</td>\n",
       "      <td>81.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_BAC_bulk_5</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>BAC</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>2</td>\n",
       "      <td>6/16/2018</td>\n",
       "      <td>15</td>\n",
       "      <td>N</td>\n",
       "      <td>169.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>KNLK_16</td>\n",
       "      <td>KNLK_16_S34</td>\n",
       "      <td>KNLK_SD132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_MF_comb_bulk_1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>MF_combined</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>18</td>\n",
       "      <td>9/8/2018</td>\n",
       "      <td>171</td>\n",
       "      <td>Y</td>\n",
       "      <td>1.070</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>20</td>\n",
       "      <td>KNLK_28</td>\n",
       "      <td>KNLK_28_S43</td>\n",
       "      <td>KNLK_SD009</td>\n",
       "      <td>43.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_MF_comb_bulk_2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>MF_combined</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>18</td>\n",
       "      <td>9/8/2018</td>\n",
       "      <td>172</td>\n",
       "      <td>Y</td>\n",
       "      <td>1.910</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>46</td>\n",
       "      <td>KNLK_51</td>\n",
       "      <td>KNLK_51_S53</td>\n",
       "      <td>KNLK_SD142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_MF_comb_bulk_3</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>MF_combined</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>18</td>\n",
       "      <td>9/8/2018</td>\n",
       "      <td>173</td>\n",
       "      <td>Y</td>\n",
       "      <td>4.850</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>50</td>\n",
       "      <td>KNLK_55</td>\n",
       "      <td>KNLK_55_S85</td>\n",
       "      <td>KNLK_SD032</td>\n",
       "      <td>85.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_biofilm_ret_scrape_1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_BF_retentate_scrapings</td>\n",
       "      <td>biofilm</td>\n",
       "      <td>Rose_Kantor</td>\n",
       "      <td>2</td>\n",
       "      <td>6/16/2018</td>\n",
       "      <td>9</td>\n",
       "      <td>N</td>\n",
       "      <td>600.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>37</td>\n",
       "      <td>KNLK_43</td>\n",
       "      <td>KNLK_43_S76</td>\n",
       "      <td>KNLK_SD024</td>\n",
       "      <td>76.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_biofilm_ret_scrape_2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_BF_retentate_scrapings</td>\n",
       "      <td>biofilm</td>\n",
       "      <td>Rose_Kantor</td>\n",
       "      <td>19</td>\n",
       "      <td>9/27/2018</td>\n",
       "      <td>199</td>\n",
       "      <td>N</td>\n",
       "      <td>30.500</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>31</td>\n",
       "      <td>KNLK_38</td>\n",
       "      <td>KNLK_38_S36</td>\n",
       "      <td>KNLK_SD006</td>\n",
       "      <td>36.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_biofilm_sep_1</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_BF_separator</td>\n",
       "      <td>biofilm</td>\n",
       "      <td>Rose_Kantor</td>\n",
       "      <td>RO1</td>\n",
       "      <td>9/13/2018</td>\n",
       "      <td>A57</td>\n",
       "      <td>Y</td>\n",
       "      <td>1.170</td>\n",
       "      <td>Powersoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>KNLK_13</td>\n",
       "      <td>KNLK_13_S10</td>\n",
       "      <td>KNLK_SD177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_biofilm_sep_2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_BF_separator</td>\n",
       "      <td>biofilm</td>\n",
       "      <td>Rose_Kantor</td>\n",
       "      <td>19</td>\n",
       "      <td>9/27/2018</td>\n",
       "      <td>A69</td>\n",
       "      <td>N</td>\n",
       "      <td>0.071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>53</td>\n",
       "      <td>KNLK_58</td>\n",
       "      <td>KNLK_58_S22</td>\n",
       "      <td>KNLK_SD182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_bulk_3</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_2stage</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>RO1</td>\n",
       "      <td>9/13/2018</td>\n",
       "      <td>A51</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.345</td>\n",
       "      <td>Powersoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>78</td>\n",
       "      <td>KNLK_80</td>\n",
       "      <td>KNLK_80_S32</td>\n",
       "      <td>KNLK_SD130</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_bulk_4</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_2stage</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>Test_June2</td>\n",
       "      <td>6/2/2018</td>\n",
       "      <td>W12</td>\n",
       "      <td>supernatant</td>\n",
       "      <td>0.603</td>\n",
       "      <td>Powersoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>KNLK_23</td>\n",
       "      <td>KNLK_23_S3</td>\n",
       "      <td>KNLK_SD179</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_bulk_5</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_2stage</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>RO1</td>\n",
       "      <td>9/13/2018</td>\n",
       "      <td>A53</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.399</td>\n",
       "      <td>Powersoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40</td>\n",
       "      <td>KNLK_46</td>\n",
       "      <td>KNLK_46_S13</td>\n",
       "      <td>KNLK_SD181</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_RO2_bulk_6</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>RO_2stage</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>RO1</td>\n",
       "      <td>9/13/2018</td>\n",
       "      <td>A55</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.187</td>\n",
       "      <td>Powersoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>74</td>\n",
       "      <td>KNLK_77</td>\n",
       "      <td>KNLK_77_S8</td>\n",
       "      <td>KNLK_SD161</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_inf_bulk_2</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>WW_3ary</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>1</td>\n",
       "      <td>6/15/2018</td>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>158.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>62</td>\n",
       "      <td>KNLK_66</td>\n",
       "      <td>KNLK_66_S86</td>\n",
       "      <td>KNLK_SD033</td>\n",
       "      <td>86.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_inf_bulk_3</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>WW_3ary</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>1</td>\n",
       "      <td>6/15/2018</td>\n",
       "      <td>3</td>\n",
       "      <td>N</td>\n",
       "      <td>680.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26</td>\n",
       "      <td>KNLK_33</td>\n",
       "      <td>KNLK_33_S83</td>\n",
       "      <td>KNLK_SD030</td>\n",
       "      <td>83.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_inf_bulk_4</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>WW_3ary</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>1</td>\n",
       "      <td>6/15/2018</td>\n",
       "      <td>4</td>\n",
       "      <td>N</td>\n",
       "      <td>518.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>38</td>\n",
       "      <td>KNLK_44</td>\n",
       "      <td>KNLK_44_S84</td>\n",
       "      <td>KNLK_SD031</td>\n",
       "      <td>84.0</td>\n",
       "      <td>+</td>\n",
       "      <td>AWTP_2_inf_bulk_5</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sample</td>\n",
       "      <td>WW_3ary</td>\n",
       "      <td>DEUF</td>\n",
       "      <td>Scott_Miller</td>\n",
       "      <td>1</td>\n",
       "      <td>6/15/2018</td>\n",
       "      <td>5</td>\n",
       "      <td>N</td>\n",
       "      <td>542.000</td>\n",
       "      <td>PowerSoil Pro</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id   read_names  sample_16S  order band_in_gel_amplicons  \\\n",
       "85    KNLK_9   KNLK_9_S65  KNLK_SD147    NaN                     +   \n",
       "25   KNLK_32  KNLK_32_S75  KNLK_SD023   75.0                     +   \n",
       "73   KNLK_76  KNLK_76_S79  KNLK_SD160    NaN                     +   \n",
       "2    KNLK_11  KNLK_11_S81  KNLK_SD028   81.0                     +   \n",
       "7    KNLK_16  KNLK_16_S34  KNLK_SD132    NaN                     +   \n",
       "20   KNLK_28  KNLK_28_S43  KNLK_SD009   43.0                     +   \n",
       "46   KNLK_51  KNLK_51_S53  KNLK_SD142    NaN                     +   \n",
       "50   KNLK_55  KNLK_55_S85  KNLK_SD032   85.0                     +   \n",
       "37   KNLK_43  KNLK_43_S76  KNLK_SD024   76.0                     +   \n",
       "31   KNLK_38  KNLK_38_S36  KNLK_SD006   36.0                     +   \n",
       "4    KNLK_13  KNLK_13_S10  KNLK_SD177    NaN                     +   \n",
       "53   KNLK_58  KNLK_58_S22  KNLK_SD182    NaN                     +   \n",
       "78   KNLK_80  KNLK_80_S32  KNLK_SD130    NaN                     +   \n",
       "15   KNLK_23   KNLK_23_S3  KNLK_SD179    NaN                     +   \n",
       "40   KNLK_46  KNLK_46_S13  KNLK_SD181    NaN                     +   \n",
       "74   KNLK_77   KNLK_77_S8  KNLK_SD161    NaN                     +   \n",
       "62   KNLK_66  KNLK_66_S86  KNLK_SD033   86.0                     +   \n",
       "26   KNLK_33  KNLK_33_S83  KNLK_SD030   83.0                     +   \n",
       "38   KNLK_44  KNLK_44_S84  KNLK_SD031   84.0                     +   \n",
       "\n",
       "                sample_code_partial Sequal_prep_elute_pooled  Duplicate.  \\\n",
       "85                AWTP_2_BAC_bulk_2                        N       False   \n",
       "25                AWTP_2_BAC_bulk_3                        N       False   \n",
       "73                AWTP_2_BAC_bulk_4                        N       False   \n",
       "2                 AWTP_2_BAC_bulk_5                        N       False   \n",
       "7             AWTP_2_MF_comb_bulk_1                        N       False   \n",
       "20            AWTP_2_MF_comb_bulk_2                        N       False   \n",
       "46            AWTP_2_MF_comb_bulk_3                        N       False   \n",
       "50  AWTP_2_RO2_biofilm_ret_scrape_1                        N       False   \n",
       "37  AWTP_2_RO2_biofilm_ret_scrape_2                        N       False   \n",
       "31         AWTP_2_RO2_biofilm_sep_1                        N       False   \n",
       "4          AWTP_2_RO2_biofilm_sep_2                        N       False   \n",
       "53                AWTP_2_RO2_bulk_3                        N       False   \n",
       "78                AWTP_2_RO2_bulk_4                        N       False   \n",
       "15                AWTP_2_RO2_bulk_5                        N       False   \n",
       "40                AWTP_2_RO2_bulk_6                        N       False   \n",
       "74                AWTP_2_inf_bulk_2                        N       False   \n",
       "62                AWTP_2_inf_bulk_3                        N       False   \n",
       "26                AWTP_2_inf_bulk_4                        N       False   \n",
       "38                AWTP_2_inf_bulk_5                        N       False   \n",
       "\n",
       "   need_optimization. Has_Other_pos_sample  ... sample_or_control  \\\n",
       "85                  N                  NaN  ...            sample   \n",
       "25                  N                  NaN  ...            sample   \n",
       "73                  N                  NaN  ...            sample   \n",
       "2                   N                  NaN  ...            sample   \n",
       "7                   N                  NaN  ...            sample   \n",
       "20                  N                  NaN  ...            sample   \n",
       "46                  N                  NaN  ...            sample   \n",
       "50                  N                  NaN  ...            sample   \n",
       "37                  N                  NaN  ...            sample   \n",
       "31                  N                  NaN  ...            sample   \n",
       "4                   N                  NaN  ...            sample   \n",
       "53                  N                  NaN  ...            sample   \n",
       "78                  N                  NaN  ...            sample   \n",
       "15                  N                  NaN  ...            sample   \n",
       "40                  N                  NaN  ...            sample   \n",
       "74                  N                  NaN  ...            sample   \n",
       "62                  N                  NaN  ...            sample   \n",
       "26                  N                  NaN  ...            sample   \n",
       "38                  N                  NaN  ...            sample   \n",
       "\n",
       "                location_code sample_type  sampler_name     Ext_batch  \\\n",
       "85                        BAC        DEUF  Scott_Miller             2   \n",
       "25                        BAC        DEUF  Scott_Miller  Test_March11   \n",
       "73                        BAC        DEUF  Scott_Miller             2   \n",
       "2                         BAC        DEUF  Scott_Miller             2   \n",
       "7                 MF_combined        DEUF  Scott_Miller            18   \n",
       "20                MF_combined        DEUF  Scott_Miller            18   \n",
       "46                MF_combined        DEUF  Scott_Miller            18   \n",
       "50  RO_BF_retentate_scrapings     biofilm   Rose_Kantor             2   \n",
       "37  RO_BF_retentate_scrapings     biofilm   Rose_Kantor            19   \n",
       "31            RO_BF_separator     biofilm   Rose_Kantor           RO1   \n",
       "4             RO_BF_separator     biofilm   Rose_Kantor            19   \n",
       "53                  RO_2stage        DEUF  Scott_Miller           RO1   \n",
       "78                  RO_2stage        DEUF  Scott_Miller    Test_June2   \n",
       "15                  RO_2stage        DEUF  Scott_Miller           RO1   \n",
       "40                  RO_2stage        DEUF  Scott_Miller           RO1   \n",
       "74                    WW_3ary        DEUF  Scott_Miller             1   \n",
       "62                    WW_3ary        DEUF  Scott_Miller             1   \n",
       "26                    WW_3ary        DEUF  Scott_Miller             1   \n",
       "38                    WW_3ary        DEUF  Scott_Miller             1   \n",
       "\n",
       "    Extraction_Date  Ext_num_unique  Microconcentrated. DNA_Ext_conc_ngperuL  \\\n",
       "85        6/16/2018              13                   N               17.500   \n",
       "25        3/11/2018              W5         supernatant               27.000   \n",
       "73        6/16/2018              14                   N               85.600   \n",
       "2         6/16/2018              15                   N              169.000   \n",
       "7          9/8/2018             171                   Y                1.070   \n",
       "20         9/8/2018             172                   Y                1.910   \n",
       "46         9/8/2018             173                   Y                4.850   \n",
       "50        6/16/2018               9                   N              600.000   \n",
       "37        9/27/2018             199                   N               30.500   \n",
       "31        9/13/2018             A57                   Y                1.170   \n",
       "4         9/27/2018             A69                   N                0.071   \n",
       "53        9/13/2018             A51                   Y                0.345   \n",
       "78         6/2/2018             W12         supernatant                0.603   \n",
       "15        9/13/2018             A53                   Y                0.399   \n",
       "40        9/13/2018             A55                   Y                0.187   \n",
       "74        6/15/2018               2                   N              158.000   \n",
       "62        6/15/2018               3                   N              680.000   \n",
       "26        6/15/2018               4                   N              518.000   \n",
       "38        6/15/2018               5                   N              542.000   \n",
       "\n",
       "              kit  \n",
       "85  PowerSoil Pro  \n",
       "25            NaN  \n",
       "73  PowerSoil Pro  \n",
       "2   PowerSoil Pro  \n",
       "7   PowerSoil Pro  \n",
       "20  PowerSoil Pro  \n",
       "46  PowerSoil Pro  \n",
       "50  PowerSoil Pro  \n",
       "37  PowerSoil Pro  \n",
       "31  Powersoil Pro  \n",
       "4             NaN  \n",
       "53  Powersoil Pro  \n",
       "78  Powersoil Pro  \n",
       "15  Powersoil Pro  \n",
       "40  Powersoil Pro  \n",
       "74  PowerSoil Pro  \n",
       "62  PowerSoil Pro  \n",
       "26  PowerSoil Pro  \n",
       "38  PowerSoil Pro  \n",
       "\n",
       "[19 rows x 46 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tab[tab.project_type=='DPWF'].sort_values('sample_code_partial')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assembly post-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "assem_dir = '/data2/other/knelson/assembly'\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    with open(f'{wd}/workflows/postAssem_{group}.sh', 'w') as f:\n",
    "        for row in df.itertuples():\n",
    "            s = row.sample_id\n",
    "            assem_name = row.sample_code_partial\n",
    "            \n",
    "            # replace the fasta headers that start \"k141\" with \"sample_id\" and rename file to sample_id_contigs.fa\n",
    "            rehead = f\"sed 's/k141/{s}/g' {assem_dir}/{assem_name}/final.contigs.fa > {assem_dir}/{assem_name}/{s}_contigs.fa\"\n",
    "                        \n",
    "            # filter for only contigs â‰¥1000 bp\n",
    "            min1000 = f'pullseq -i {assem_dir}/{assem_name}/{s}_contigs.fa --min 1000 > {assem_dir}/{assem_name}/{s}_contigs_min1000.fa'\n",
    "            \n",
    "            # delete extra files from assembly\n",
    "            clean = f'rm -r {assem_dir}/{assem_name}/intermediate_contigs/ '\\\n",
    "                    f'{assem_dir}/{assem_name}/checkpoints.txt '\\\n",
    "                    f'{assem_dir}/{assem_name}/final.contigs.fa '\\\n",
    "                    f'{assem_dir}/{assem_name}/done '\\\n",
    "                    f'{assem_dir}/{assem_name}/options.json'\n",
    "           \n",
    "            # make directory to store bowtie2 indices in\n",
    "            mdbt2 = f'mkdir {assem_dir}/{assem_name}/bt2/'\n",
    "            \n",
    "            # index in prep for bowtie2 mapping\n",
    "            ind = f'bowtie2-build {assem_dir}/{assem_name}/{s}_contigs_min1000.fa {assem_dir}/{assem_name}/bt2/{s}_contigs_min1000.fa'\n",
    "           \n",
    "            cmd = [rehead, min1000, clean, mdbt2, ind]\n",
    "            f.write('\\n'.join(cmd) + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read-mapping across samples of same location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "bt2 = '/opt/bin/bio/bowtie2'\n",
    "ssam = '/opt/bin/bio/shrinksam'\n",
    "assem_dir = '/data2/other/knelson/assembly'\n",
    "reads_dir = '/data2/other/knelson/raw.d/trimmed_reads'\n",
    "\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    \n",
    "    # commands for each project saved separately\n",
    "    with open(f'{wd}/workflows/mapping_{group}.sh', 'w') as f:\n",
    "        \n",
    "        # group by location and cross-map within location (these reads are most likely to map and will help with binning)\n",
    "        for location, dfl in df.groupby('location_code'):\n",
    "            \n",
    "            #commands for each sample within the project\n",
    "            for row in dfl.itertuples():\n",
    "                s = row.sample_id\n",
    "                assem_name = row.sample_code_partial\n",
    "\n",
    "                for r in dfl.sample_id:\n",
    "                    out = f'{assem_dir}/{assem_name}/{s}_contigs_min1000.fa-vs-{r}.sam'\n",
    "                    bt2ind = f'{assem_dir}/{assem_name}/bt2/{s}_contigs_min1000.fa'\n",
    "                    r1 = f'{reads_dir}/{r}.PE.1.fastq.gz'\n",
    "                    r2 = f'{reads_dir}/{r}.PE.2.fastq.gz'\n",
    "                    map_cmd = f'{bt2} -p 48 -x {bt2ind} -1 {r1} -2 {r2} --reorder | {ssam} -v > {out}'\n",
    "                    qcmd = f'echo \"{map_cmd}\" | qsub -V -pe smp 48 -N {s}_vs_{r}'\n",
    "                    f.write(qcmd + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter, compress, and index mapping files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### NOT FINISHED ####\n",
    "\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    with open(f'{wd}/workflows/mapping_{group}.sh', 'w') as f:\n",
    "        f'{assem_dir}/{assem_name}/{s}_contigs_min1000.fa-vs-{r}.sam'\n",
    "        samtools view -F 4 -bS $i\"_scaffold_min2500.vs.\"{s}.sam > $i\"_scaffold_min2500.vs.\"$s.bam\n",
    "        samtools sort -m 5G $i\"_scaffold_min2500.vs.\"$s.bam > $i\"_scaffold_min2500.vs.\"$s.sorted.bam\n",
    "        samtools index $i\"_scaffold_min2500.vs.\"$s.sorted.bam\n",
    "        rm $i\"_scaffold_min2500.vs.\"$s.sam\n",
    "        rm $i\"_scaffold_min2500.vs.\"$s.bam\n",
    "        \n",
    "####        \n",
    "assem_dir = '/data2/other/knelson/assembly'\n",
    "\n",
    "for group, df in tab.groupby('project_type'):\n",
    "    \n",
    "    # commands for each project saved separately\n",
    "    with open(f'{wd}/workflows/processMapping_{group}.sh', 'w') as f:\n",
    "        \n",
    "        # group by location and cross-map within location (these reads are most likely to map and will help with binning)\n",
    "        for location, dfl in df.groupby('location_code'):\n",
    "            \n",
    "            #commands for each sample within the project\n",
    "            for row in dfl.itertuples():\n",
    "                s = row.sample_id\n",
    "                assem_name = row.sample_code_partial\n",
    "\n",
    "                for r in dfl.sample_id:\n",
    "                    out = f'{assem_dir}/{assem_name}/{s}_contigs_min1000.fa-vs-{r}.sam'\n",
    "                    bt2ind = f'{assem_dir}/{assem_name}/bt2/{s}_contigs_min1000.fa'\n",
    "                    r1 = f'{reads_dir}/{r}.PE.1.fastq.gz'\n",
    "                    r2 = f'{reads_dir}/{r}.PE.2.fastq.gz'\n",
    "                    map_cmd = f'{bt2} -p 48 -x {bt2ind} -1 {r1} -2 {r2} --reorder | {ssam} -v > {out}'\n",
    "                    qcmd = f'echo \"{map_cmd}\" | qsub -V -pe smp 48 -N {s}_vs_{r}'\n",
    "                    f.write(qcmd + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Anvi'o profiling"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
